{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The case of diatoms, dinoflagellates, and Prymnesiophytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.linear_model import LinearRegression, LogisticRegression\n",
    "from sklearn.model_selection import cross_val_score, cross_validate\n",
    "from sklearn.preprocessing import PolynomialFeatures\n",
    "from sklearn.pipeline import make_pipeline\n",
    "import matplotlib.colors as mcolors\n",
    "import plotly.graph_objects as go\n",
    "from scipy import stats\n",
    "from scipy.optimize import curve_fit\n",
    "from sklearn.metrics import r2_score, mean_squared_error\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "from statsmodels.stats.multicomp import pairwise_tukeyhsd\n",
    "\n",
    "sns.set_style('white')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By correlating functional modules with taxonomic classes, we see that on an ecosystem level, phenology drives the metabolic functions that are found. This pattern potentially obscures the relationships between species occurrence and functional diversity. \n",
    "However, there's two groups that are particularly interesting: diatoms and dinoflagellates. Both are present year-round, however a turnover in relative abundance and succession of different assemblages typifies the diatom community.\n",
    "Our metatranscriptomic data thus allows us to answer key ecological questions as a case study:\n",
    "1. What is the relation between the amount of species inhabiting a same niche and functional diversity?\n",
    "2. Is there a link between evenness and functional diversity?\n",
    "3. Do different assemblages have different functional profiles, and if so, what are the differences?\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Load data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Gather taxonomic data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we'll look at the taxonomically annotated data. In order to capture all transcripts remotely identified as diatoms or dinoflagellates, all transcripts that are aligned to either of these groups with a percent identity of 60% or higher are included. This is a very loose threshold, but it allows us to capture all transcripts that are remotely related to these groups. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load PhyloDB alignments\n",
    "data_tax = pd.read_table('../../data/annotation/taxonomy_phyloDB_extended/phylodb_extended.firsthit.90plus_alnscore.m8',  engine='pyarrow', header=None)\n",
    "print(f'The mmseqs alignment has {len(data_tax)} hits with a >60 % identity')\n",
    "data_tax.columns = ['query_id', 'target_id', 'p_ident', 'alnlen', 'mismatch', 'gapopen', 'qstart', 'qend', 'tstart', 'tend', 'evalue', 'bits']\n",
    "print(f'The mmseqs alignment contains {len(data_tax[\"query_id\"].unique())} unique query sequences')\n",
    "# Fix transcript names in the first column so that they equal the transcript identifiers in the count files\n",
    "data_tax.iloc[:, 0] = data_tax.iloc[:, 0].str.split(\".\", expand=True).drop(columns=1)\n",
    "print(f'The mmseqs alignment contains {len(data_tax[\"query_id\"].unique())} unique query sequences')\n",
    "\n",
    "# Add taxonomic information\n",
    "taxonomy = pd.read_table('../../data/annotation/taxonomy_phyloDB_extended/phylodb_1.076.taxonomy_extended.txt')\n",
    "print(f'The taxonomy file contains {len(taxonomy)} rows')\n",
    "annotation = pd.read_table('../../data/annotation/taxonomy_phyloDB_extended/phylodb_1.076.annotations_extended.txt', engine='pyarrow', header=None)\n",
    "print(f'The annotation file contains {len(annotation)} rows, linking transcript IDs to their taxonomy and function')\n",
    "annotation.columns = ['target_id', 'code', 'strain_name', 'function']\n",
    "\n",
    "# Add the taxonomy information to the data_tax\n",
    "# data_tax = data_tax.merge(annotation, left_on='target_id', right_on='target_id', how='left')\n",
    "data_tax = data_tax.merge(annotation, left_on='target_id', right_on='target_id')\n",
    "print(f'The alignment merged with annotation file contains {len(data_tax)} rows')\n",
    "\n",
    "data_tax = data_tax.merge(taxonomy, left_on='strain_name', right_on = 'strain_name')\n",
    "print(f'The data_tax merged with the taxonomy  file contains {len(data_tax)} rows')\n",
    "\n",
    "data_tax = data_tax.drop(columns=['code', 'peptide_count', 'mismatch', 'gapopen', 'qstart', 'qend', 'tstart', 'tend', 'evalue', 'bits'])\n",
    "# Expand taxonomy path in relevant columns\n",
    "data_tax[['kingdom', 'superphylum', 'phylum', 'class', 'order', 'family', 'genus', 'species']] = data_tax['taxonomy'].str.split(';', expand = True)\n",
    "\n",
    "# View data_tax\n",
    "data_tax.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_tax.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1.2 Alternative taxonomy: EukProt\n",
    "Alternatively, do this analysis with the EukProt taxonomy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Load eukprot annotations\n",
    "data_tax_eukprot = pd.read_table('../../data/annotation/taxonomy_eukprot/eukprot_DB.firsthit.60plus_alnscore.m8', header=None)\n",
    "print(f'The eukprot annotation file contains {len(data_tax_eukprot)} rows')\n",
    "# Fix transcript names in the first column so that they equal the transcript identifiers in the count files\n",
    "data_tax_eukprot.iloc[:, 0] = data_tax_eukprot.iloc[:, 0].str.split(\".\", expand=True).drop(columns=1)\n",
    "\n",
    "## In the second column, split of the EukProt ID off\n",
    "eukprot_ID = data_tax_eukprot.iloc[:, 1].str.split(\"_\", expand=True)[0]\n",
    "data_tax_eukprot.iloc[:, 1] = eukprot_ID\n",
    "data_tax_eukprot.columns = ['query_id', 'target_id', 'p_ident', 'alnlen', 'mismatch', 'gapopen', 'qstart', 'qend', 'tstart', 'tend', 'evalue', 'bits']\n",
    "\n",
    "## Add taxonomic information\n",
    "eukprot_taxonomy = pd.read_table('../../data/annotation/taxonomy_eukprot/EukProt_included_data_sets.v03.2021_11_22.txt')\n",
    "print(f'The eukprot taxonomy file contains {len(eukprot_taxonomy)} rows')\n",
    "\n",
    "# Drop the columns that are not needed\n",
    "eukprot_taxonomy.drop(columns=['Previous_Names', 'Replaces_EukProt_ID', 'Data_Source_URL', 'Data_Source_Name', 'Paper_DOI', 'Actions_Prior_to_Use',\n",
    "       'Data_Source_Type', 'Notes', 'Columns_Modified_Since_Previous_Version', 'Merged_Strains',\n",
    "       'Alternative_Strain_Names', '18S_Sequence_GenBank_ID', '18S_Sequence',\n",
    "       '18S_Sequence_Source', '18S_Sequence_Other_Strain_GenBank_ID',\n",
    "       '18S_Sequence_Other_Strain_Name', '18S_and_Taxonomy_Notes'], inplace=True)\n",
    "\n",
    "# Swap the _ to a space in the Name_to_Use column\n",
    "eukprot_taxonomy['Name_to_Use'] = eukprot_taxonomy['Name_to_Use'].str.replace('_', ' ')\n",
    "\n",
    "# Merge the annotation and taxonomy files\n",
    "data_tax_eukprot = data_tax_eukprot.merge(eukprot_taxonomy, left_on='target_id', right_on='EukProt_ID', how='left')\n",
    "print(f'The eukprot annotation merged with taxonomy file contains {len(data_tax_eukprot)} rows')\n",
    "\n",
    "# Drop the columns that are not needed\n",
    "data_tax_eukprot.drop(columns=['target_id', 'gapopen', 'qstart', 'qend', 'tstart', 'tend', 'evalue', 'bits'], inplace=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Gather functional data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We'll also look at the functional annotations, which are based on the KEGG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the annotation data\n",
    "data_func = pd.read_csv('../../data/annotation/functional_eggnog/functional_annotation.emapper.annotations', sep = '\\t', engine = 'pyarrow')\n",
    "\n",
    "# Fix transcript names in the first column so that they equal the transcript identifiers in the count files\n",
    "# This is necessary because TransDecoder adds .p2 or .p1 to the sequence identifiers\n",
    "data_func.iloc[:, 0] = data_func.iloc[:, 0].str.split(\".\", expand=True).drop(columns=1)\n",
    "\n",
    "# Check the annotation data, how many more hits are there compared to unique query sequences?\n",
    "print(f'number of rows in eggNOG annotation: {len(data_func)}')\n",
    "print(f'number of unique query sequences in eggNOG annotation: {len(data_func.iloc[:, 0].unique())}')\n",
    "\n",
    "# Check all columns in the annotation data\n",
    "data_func.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_func.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only retain relevant columns and rename them\n",
    "data_func = data_func[['#query',\n",
    "             'Description', 'GOs', 'KEGG_ko', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction', 'PFAMs']]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 Gather count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the necessary count data\n",
    "counts = pd.read_csv('../../data/kallisto/tpm.csv', engine='pyarrow')\n",
    "## This is the transcript read mapping, prior to protein prediction!\n",
    "# Rename the first column to transcript_id\n",
    "counts = counts.rename(columns={'target_id': 'transcript_id'})\n",
    "print(f'{len(counts)} transcripts were quantified (on the nucleotide level) in the kallisto run')\n",
    "# Transform the data to the long format\n",
    "counts = counts.melt(id_vars=['transcript_id'],var_name='sample', value_name='TPM')\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3.2 Gather TPL daata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the necessary count data\n",
    "tpl = pd.read_csv('../../data/kallisto/transcripts_per_L.csv', engine='pyarrow')\n",
    "## This is the transcript read mapping, prior to protein prediction!\n",
    "# Rename the first column to transcript_id\n",
    "tpl = tpl.rename(columns={'target_id': 'transcript_id'})\n",
    "# Transform the data to the long format\n",
    "tpl = tpl.melt(id_vars=['transcript_id'],var_name='sample', value_name='TPL')\n",
    "tpl.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 Combine data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Combine count and annotation data together\n",
    "## Prefilter\n",
    "# Combine relevant taxonomic and functional data\n",
    "#data = data[(data['class'] == 'Bacillariophyta') | (data['class'] == 'Dinophyceae') | (data['class'] == 'Prymnesiophyceae')]\n",
    "data_tax_eukprot = data_tax_eukprot[(data_tax_eukprot['Taxogroup1_UniEuk'] == 'Ochrophyta') | (data_tax_eukprot['Taxogroup1_UniEuk'] == 'Dinoflagellata') | (data_tax_eukprot['Taxogroup1_UniEuk'] == 'Prymnesiophyceae')]\n",
    "\n",
    "# Merge with PhyloDB or EUKprot, TPM or TPL data - depending on which code is run downstream.\n",
    "#data = data_tax.merge(counts, left_on='query_id', right_on='transcript_id', how='left')\n",
    "#data = data_tax_eukprot.merge(counts, left_on='query_id', right_on='transcript_id', how='left')\n",
    "#del data_tax_eukprot, counts\n",
    "data = data_tax_eukprot.merge(tpl, left_on='query_id', right_on='transcript_id', how='left')\n",
    "del tpl\n",
    "\n",
    "data = data.drop(columns='query_id')\n",
    "\n",
    "# Add sample metadata\n",
    "meta = pd.read_csv('../../samples.csv', sep=';', index_col=0)\n",
    "data = data.merge(meta, left_on= 'sample', right_on= 'sample', how = 'left')\n",
    "\n",
    "# View data\n",
    "print(f'{len(data)} rows are in the data, which should be the amount of annotated proteins times the number of samples')\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#data[['class', 'TPM']].groupby('class').sum().sort_values(by='TPM', ascending=False)\n",
    "data[['Taxogroup1_UniEuk', 'TPM']].groupby('Taxogroup1_UniEuk').sum().sort_values(by='TPM', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#data[['class', 'TPM']].groupby('class').sum().sort_values(by='TPM', ascending=False)\n",
    "data[['Taxogroup1_UniEuk', 'TPL']].groupby('Taxogroup1_UniEuk').sum().sort_values(by='TPM', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#data[['class', 'TPM']].groupby('class').sum().sort_values(by='TPM', ascending=False)\n",
    "data[['Taxogroup2_UniEuk', 'TPM']].groupby('Taxogroup2_UniEuk').sum().sort_values(by='TPM', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.merge(data_func, left_on='transcript_id', right_on='#query')\n",
    "data = data.drop(columns=['#query'])\n",
    "print(len(data))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Diatoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, let's get a list of the unique species in the data\n",
    "#data[(data['class'] == 'Bacillariophyta') & (data['p_ident'] >= 0.98)]['species'].unique()\n",
    "data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.98)]['Name_to_Use'].unique()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Diatom Abundance\n",
    "First we'll make the graph showing relative diatom abundances per month throughout the year. After, we'll show absolute diatom abundances per month."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the data\n",
    "## Get the diatoms\n",
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "\n",
    "## Since we'll be looking at the relative abundance of different diatom genera, we can only include reads that are annotated to a genus with a sufficient % sequence identity\n",
    "#data_diatoms = data[(data['class'] == 'Bacillariophyta') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_diatoms = data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_diatoms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove transcripts below a certain TPM threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPM'] > 1]\n",
    "\n",
    "# Group by month and genus, sum TPM\n",
    "data_diatoms = data_diatoms.groupby([aggregation_level, tax_level]).sum().reset_index()\n",
    "\n",
    "data_diatoms['month'] = pd.Categorical(data_diatoms['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Normalise mean of TPM values to the total mean TPM of that month\n",
    "data_diatoms[\"rel_expression_per_month\"] = data_diatoms.TPM / data_diatoms.groupby('month').TPM.transform('sum')\n",
    "\n",
    "# Combine low-abundant groups\n",
    "rare_groups = data_diatoms['rel_expression_per_month'] <= 0.02\n",
    "data_diatoms.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "# Print unique species\n",
    "print(data_diatoms[tax_level].unique())\n",
    "# Inspect data\n",
    "data_diatoms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "## First, set conversion factor to transform pixels to cm\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "fig = px.histogram(data_diatoms.sort_values(\"month\", ascending=False), \n",
    "                x=\"rel_expression_per_month\", \n",
    "                y=\"month\", \n",
    "                color=tax_level,\n",
    "                orientation='h',\n",
    "                color_discrete_map=(\n",
    "                    {'Rare': '#808080', \n",
    "                     'Asterionellopsis': '#ff8c00', \n",
    "                     'Astrosyne': '#87cefa',\n",
    "                     'Chaetoceros': '#E69F00',\n",
    "                     'Corethron': '#ffd700',\n",
    "                     'Coscinodiscus': '#B5B5B5',\n",
    "                     'Craspedostauros': '#2196F3',\n",
    "                     'Dactyliosolen': '#ba55d3',\n",
    "                     'Ditylum': '#56B4E9',\n",
    "                     'Entomoneis': '#795548',\n",
    "                     'Eucampia': '#89CE00',\n",
    "                     'Guinardia': '#009E73', \n",
    "                     'Helicotheca': '#04686E',\n",
    "                     'Leptocylindrus': '#046E0A',\n",
    "                     'Odontella': '#0072B2',\n",
    "                     'Pseudo-nitzschia': '#A42324',\n",
    "                     'Rhizosolenia': '#CC79A7',\n",
    "                     'Skeletonema': '#fa8072',\n",
    "                     'Stephanopyxis': '#dda0dd',\n",
    "                     'Thalassiosira': '#D55E00',\n",
    "                     'Triceratium': '#CC7DFF',\n",
    "                     'Amphiprora': '#c2ccc4' \n",
    "                     }\n",
    "                ),\n",
    "                #color_discrete_map={\n",
    "                #     'Rare': '#808080',\n",
    "                #     'Asterionellopsis': '#ff8c00',\n",
    "                #     'Chaetoceros': '#E69F00',\n",
    "                #     'Cyclotella': '#8BC34A',\n",
    "                #     'Dactyliosolen': '#ba55d3',\n",
    "                #     'Ditylum': '#56B4E9',\n",
    "                #     'Helicotheca': '#04686E',\n",
    "                #     'Pseudictyota': '#673AB7',\n",
    "                #     'Pseudo-nitzschia': '#A42324',\n",
    "                #     'Rhizosolenia': '#CC79A7',\n",
    "                #     'Thalassiosira': '#D55E00',\n",
    "                #     'Trieres': '#F8EF3A',\n",
    "                #     'Astrosyne': '#87cefa',\n",
    "                #     'Coscinodiscus': '#B5B5B5',\n",
    "                #     'Entomoneis': '#795548',\n",
    "                #     'Extubocellulus': '#607D8B',\n",
    "                #     'Gedaniella': '#3F51B5',\n",
    "                #     'Leptocylindrus': '#046E0A',\n",
    "             #     'Phaeodactylum': '#9E9E9E',\n",
    "                #     'Odontella': '#0072B2',\n",
    "                #     'Corethron': '#ffd700',\n",
    "                #     'Licmophora': '#98fb98',\n",
    "                #     'Proboscia': '#0000ff',\n",
    "                #     'Stephanopyxis': '#dda0dd',\n",
    "                #     'Synedropsis': '#673AB7',\n",
    "                #     'Aulacoseira': '#00BCD4',\n",
    "                #     'Skeletonema': '#fa8072',\n",
    "                #     'Eucampia': '#89CE00',\n",
    "                #     'Fragilariopsis': '#FFEB3B',\n",
    "                #     'Attheya': '#795548',\n",
    "                #     'Thalassionema': '#9C27B0',\n",
    "                #     'Craspedostauros': '#2196F3'\n",
    "                # },\n",
    "                # Specify all the months that need to be included, \n",
    "                # even if no sample has been taken\n",
    "                category_orders={\"genus\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                            'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                            'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', \n",
    "                                            'Entomoneis', 'Eucampia', 'Craspedostauros', 'Amphiprora', 'Guinardia', 'Triceratium'],\n",
    "                                 #\"Genus_UniEuk\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                 #                'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                 #                'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', 'Licmophora']\n",
    "                    }\n",
    "                # text_auto='.2f'\n",
    "                )\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='% TPM of total sum',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save figure as png\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_phylodb.png\".format(tax_level), scale=1)\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_phylodb.svg\".format(tax_level), scale=1)\n",
    "\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_eukprot.png\".format(tax_level), scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_eukprot.svg\".format(tax_level), scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store the order of the genera in the legend\n",
    "diatom_legend_order = []\n",
    "for i in range(len(fig.data)):\n",
    "    diatom_legend_order.append(fig.data[i].name)\n",
    "print(diatom_legend_order)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Spatial distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "#data_diatoms = data[(data['class'] == 'Bacillariophyta') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "data_diatoms = data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPM'] > 1]\n",
    "\n",
    "# Group by month, station and species, sum TPM\n",
    "data_diatoms = data_diatoms.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "data_diatoms['month'] = pd.Categorical(data_diatoms['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Create list of genera for color assignment\n",
    "diatom_genera = []\n",
    "# Plot per station\n",
    "for station in data_diatoms[aggregation_level2].unique():\n",
    "    plot_data = data_diatoms[data_diatoms[aggregation_level2] == station]\n",
    "    # Normalise mean of TPM values to the total mean TPM of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPM / plot_data.groupby('month').TPM.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "    # Add all unique genera to dictionary if not already present\n",
    "    for genus in plot_data[tax_level].unique():\n",
    "        diatom_genera.append(genus) if genus not in diatom_genera else None\n",
    "    # Plot\n",
    "    fig = px.histogram(plot_data.sort_values(\"month\", ascending=False), \n",
    "                    x=\"rel_expression_per_month\", \n",
    "                    y=\"month\", \n",
    "                    color=tax_level,\n",
    "                    orientation='h',\n",
    "                    # text_auto='.2f',\n",
    "                    color_discrete_map=(\n",
    "                        {'Rare': '#808080', \n",
    "                         'Asterionellopsis': '#ff8c00', \n",
    "                         'Astrosyne': '#87cefa',\n",
    "                         'Chaetoceros': '#E69F00',\n",
    "                         'Corethron': '#ffd700',\n",
    "                         'Coscinodiscus': '#B5B5B5',\n",
    "                         'Craspedostauros': '#2196F3',\n",
    "                         'Dactyliosolen': '#ba55d3',\n",
    "                         'Ditylum': '#56B4E9',\n",
    "                         'Entomoneis': '#795548',\n",
    "                         'Eucampia': '#89CE00',\n",
    "                         'Guinardia': '#009E73', \n",
    "                         'Helicotheca': '#04686E',\n",
    "                         'Leptocylindrus': '#046E0A',\n",
    "                         'Odontella': '#0072B2',\n",
    "                         'Pseudo-nitzschia': '#A42324',\n",
    "                         'Rhizosolenia': '#CC79A7',\n",
    "                         'Skeletonema': '#fa8072',\n",
    "                         'Stephanopyxis': '#dda0dd',\n",
    "                         'Thalassiosira': '#D55E00',\n",
    "                         'Triceratium': '#CC7DFF',\n",
    "                         'Amphiprora': '#c2ccc4' \n",
    "                         }\n",
    "                    ),\n",
    "                    #color_discrete_map={\n",
    "                    #    'Rare': '#808080',\n",
    "                    #    'Asterionellopsis': '#ff8c00',\n",
    "                    #    'Chaetoceros': '#E69F00',\n",
    "                    #    'Cyclotella': '#8BC34A',\n",
    "                    #    'Dactyliosolen': '#ba55d3',\n",
    "                    #    'Ditylum': '#56B4E9',\n",
    "                    #    'Helicotheca': '#04686E',\n",
    "                    #    'Pseudictyota': '#673AB7',\n",
    "                    #    'Pseudo-nitzschia': '#A42324',\n",
    "                    #    'Rhizosolenia': '#CC79A7',\n",
    "                    #    'Thalassiosira': '#D55E00',\n",
    "                    #    'Trieres': '#F8EF3A',\n",
    "                    #    'Astrosyne': '#87cefa',\n",
    "                    #    'Coscinodiscus': '#B5B5B5',\n",
    "                    #    'Entomoneis': '#795548',\n",
    "                    #    'Extubocellulus': '#607D8B',\n",
    "                    #    'Gedaniella': '#3F51B5',\n",
    "                    #    'Leptocylindrus': '#046E0A',\n",
    "                    #    'Phaeodactylum': '#9E9E9E',\n",
    "                    #    'Odontella': '#0072B2',\n",
    "                    #    'Corethron': '#ffd700',\n",
    "                    #    'Licmophora': '#98fb98',\n",
    "                    #    'Proboscia': '#0000ff',\n",
    "                    #    'Stephanopyxis': '#dda0dd',\n",
    "                    #    'Synedropsis': '#673AB7',\n",
    "                    #    'Aulacoseira': '#00BCD4',\n",
    "                    #    'Skeletonema': '#fa8072',\n",
    "                    #    'Eucampia': '#89CE00',\n",
    "                    #    'Fragilariopsis': '#FFEB3B',\n",
    "                    #    'Attheya': '#795548',\n",
    "                    #    'Thalassionema': '#9C27B0',\n",
    "                    #    'Craspedostauros': '#2196F3'\n",
    "                    #},\n",
    "\n",
    "                    # Specify all the months that need to be included, \n",
    "                    # even if no sample has been taken\n",
    "                    category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                                        \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                                        \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                                        \"June_2021\", \"July_2021\"],\n",
    "                                     \"genus\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                            'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                            'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', \n",
    "                                            'Entomoneis', 'Eucampia', 'Craspedostauros', 'Amphiprora', 'Guinardia', 'Triceratium'],\n",
    "                                      #\"Genus_UniEuk\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                      #               'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                      #               'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', 'Licmophora']\n",
    "\n",
    "                        }\n",
    "                    )\n",
    "\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "            size=8,  # Set the font size\n",
    "            color=\"#000000\"  # Set the font color\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width= 8.5 * pixels_per_cm,\n",
    "        height= 7.5 * pixels_per_cm,\n",
    "        margin=dict( # Set the margins\n",
    "            l=0,  # Left margin\n",
    "            r=25,  # Right margin\n",
    "            b=25,  # Bottom margin\n",
    "            t=25  # Top margin\n",
    "        ),\n",
    "        xaxis_title_text='% TPM of total sum',\n",
    "        yaxis_title_text=None,\n",
    "        ## Add station name to title\n",
    "        title_text=station\n",
    "    )\n",
    "\n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as png\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_at_{}_eukprot_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_expression_per_month_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diatom_genera\n",
    "# Now create colours for all of these in the above code, set the order, and rerun the above code"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Total TPM of diatoms per month"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the total TPM per month per genus\n",
    "# Normalise mean of TPM values to the total mean TPM of that month\n",
    "data_diatoms[\"rel_expression_per_month\"] = data_diatoms.TPM / data_diatoms.groupby('month').TPM.transform('sum')\n",
    "\n",
    "# Combine low-abundant groups\n",
    "rare_groups = data_diatoms['rel_expression_per_month'] <= 0.02\n",
    "data_diatoms.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "fig = px.histogram(data_diatoms.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = tax_level,\n",
    "             # Pick the colors for the stations\n",
    "            color_discrete_map=(\n",
    "                {'Rare': '#808080', \n",
    "                 'Asterionellopsis': '#ff8c00', \n",
    "                 'Astrosyne': '#87cefa',\n",
    "                 'Chaetoceros': '#E69F00',\n",
    "                 'Corethron': '#ffd700',\n",
    "                 'Coscinodiscus': '#B5B5B5',\n",
    "                 'Craspedostauros': '#2196F3',\n",
    "                 'Dactyliosolen': '#ba55d3',\n",
    "                 'Ditylum': '#56B4E9',\n",
    "                 'Entomoneis': '#795548',\n",
    "                 'Eucampia': '#89CE00',\n",
    "                 'Guinardia': '#009E73', \n",
    "                 'Helicotheca': '#04686E',\n",
    "                 'Leptocylindrus': '#046E0A',\n",
    "                 'Odontella': '#0072B2',\n",
    "                 'Pseudo-nitzschia': '#A42324',\n",
    "                 'Rhizosolenia': '#CC79A7',\n",
    "                 'Skeletonema': '#fa8072',\n",
    "                 'Stephanopyxis': '#dda0dd',\n",
    "                 'Thalassiosira': '#D55E00',\n",
    "                 'Triceratium': '#CC7DFF',\n",
    "                 'Amphiprora': '#c2ccc4' \n",
    "                 }\n",
    "            ),\n",
    "             #color_discrete_map={\n",
    "             #        'Rare': '#808080',\n",
    "             #        'Asterionellopsis': '#ff8c00',\n",
    "             #        'Chaetoceros': '#E69F00',\n",
    "             #        'Cyclotella': '#8BC34A',\n",
    "             #        'Dactyliosolen': '#ba55d3',\n",
    "             #        'Ditylum': '#56B4E9',\n",
    "             #        'Helicotheca': '#04686E',\n",
    "             #        'Pseudictyota': '#673AB7',\n",
    "             #        'Pseudo-nitzschia': '#A42324',\n",
    "             #        'Rhizosolenia': '#CC79A7',\n",
    "             #        'Thalassiosira': '#D55E00',\n",
    "             #        'Trieres': '#F8EF3A',\n",
    "             #        'Astrosyne': '#87cefa',\n",
    "             #        'Coscinodiscus': '#B5B5B5',\n",
    "             #        'Entomoneis': '#795548',\n",
    "             #        'Extubocellulus': '#607D8B',\n",
    "             #        'Gedaniella': '#3F51B5',\n",
    "             #        'Leptocylindrus': '#046E0A',\n",
    "             #        'Phaeodactylum': '#9E9E9E',\n",
    "             #        'Odontella': '#0072B2',\n",
    "             #        'Corethron': '#ffd700',\n",
    "             #        'Licmophora': '#98fb98',\n",
    "             #        'Proboscia': '#0000ff',\n",
    "             #        'Stephanopyxis': '#dda0dd',\n",
    "             #        'Synedropsis': '#673AB7',\n",
    "             #        'Aulacoseira': '#00BCD4',\n",
    "             #        'Skeletonema': '#fa8072',\n",
    "             #        'Eucampia': '#89CE00',\n",
    "             #        'Fragilariopsis': '#FFEB3B',\n",
    "             #        'Attheya': '#795548',\n",
    "             #        'Thalassionema': '#9C27B0',\n",
    "             #        'Craspedostauros': '#2196F3'\n",
    "             #    },\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"],\n",
    "                \"genus\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                            'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                            'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', \n",
    "                                            'Entomoneis', 'Eucampia', 'Craspedostauros', 'Amphiprora', 'Guinardia', 'Triceratium'],\n",
    "                'Genus_UniEuk': ['Rare', 'Helicotheca', 'Rhizosolenia', 'Leptocylindrus', 'Odontella', 'Pseudictyota',\n",
    "                                 'Trieres', 'Pseudo-nitzschia', 'Thalassiosira', 'Licmophora', 'Asterionellopsis', 'Astrosyne', \n",
    "                                 'Coscinodiscus', 'Chaetoceros', 'Entomoneis', 'Extubocellulus', 'Dactyliosolen', 'Corethron', \n",
    "                                 'Stephanopyxis', 'Proboscia', 'Ditylum', 'Skeletonema', 'Eucampia']\n",
    "                })\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total diatom TPM per month',\n",
    "    yaxis_title_text=None,\n",
    "    # Set range of x-axis\n",
    "    xaxis_range=[0, 650000],\n",
    ")\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_{tax_level}_phylodb.svg\", scale=1)\n",
    "fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_{tax_level}_eukprot.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the total TPM per month per station\n",
    "fig = px.histogram(data_diatoms.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = 'station',\n",
    "             # Pick the colors for the stations\n",
    "            color_discrete_map={\n",
    "                    \"ZG02\": \"#8c613c\",\n",
    "                    \"120\": \"#956cb4\",\n",
    "                    \"330\": \"#ee854a\",\n",
    "                    \"130\": \"#4878d0\",\n",
    "                    \"780\": \"#d65f5f\",\n",
    "                    \"700\": \"#6acc64\"},\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"]})\n",
    "\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total diatom TPM per month',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_station_phylodb.svg\", scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_station_eukprot.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Spatial distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "#data_diatoms = data[(data['class'] == 'Bacillariophyta') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "data_diatoms = data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPM'] > 1]\n",
    "\n",
    "# Group by month, station and species, sum TPM\n",
    "data_diatoms = data_diatoms.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "# Specify the desired order of months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "               \"February_2021\", \"April_2021\", \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Plot per station\n",
    "for station in data_diatoms[aggregation_level2].unique():\n",
    "    plot_data = data_diatoms[data_diatoms[aggregation_level2] == station]\n",
    "    # Normalise mean of TPM values to the total mean TPM of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPM / plot_data.groupby('month').TPM.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "    fig = px.histogram(plot_data, x = 'TPM', y=aggregation_level, color = tax_level,\n",
    "                 # Pick the colors for the stations\n",
    "                color_discrete_map=(\n",
    "                    {'Rare': '#808080', \n",
    "                     'Asterionellopsis': '#ff8c00', \n",
    "                     'Astrosyne': '#87cefa',\n",
    "                     'Chaetoceros': '#E69F00',\n",
    "                     'Corethron': '#ffd700',\n",
    "                     'Coscinodiscus': '#B5B5B5',\n",
    "                     'Craspedostauros': '#2196F3',\n",
    "                     'Dactyliosolen': '#ba55d3',\n",
    "                     'Ditylum': '#56B4E9',\n",
    "                     'Entomoneis': '#795548',\n",
    "                     'Eucampia': '#89CE00',\n",
    "                     'Guinardia': '#009E73', \n",
    "                     'Helicotheca': '#04686E',\n",
    "                     'Leptocylindrus': '#046E0A',\n",
    "                     'Odontella': '#0072B2',\n",
    "                     'Pseudo-nitzschia': '#A42324',\n",
    "                     'Rhizosolenia': '#CC79A7',\n",
    "                     'Skeletonema': '#fa8072',\n",
    "                     'Stephanopyxis': '#dda0dd',\n",
    "                     'Thalassiosira': '#D55E00',\n",
    "                     'Triceratium': '#CC7DFF',\n",
    "                     'Amphiprora': '#c2ccc4' \n",
    "                     }\n",
    "                ),\n",
    "                #color_discrete_map={\n",
    "                #        'Rare': '#808080',\n",
    "                #        'Asterionellopsis': '#ff8c00',\n",
    "                #        'Chaetoceros': '#E69F00',\n",
    "                #        'Cyclotella': '#8BC34A',\n",
    "                #        'Dactyliosolen': '#ba55d3',\n",
    "                #        'Ditylum': '#56B4E9',\n",
    "                #        'Helicotheca': '#04686E',\n",
    "                #        'Pseudictyota': '#673AB7',\n",
    "                #        'Pseudo-nitzschia': '#A42324',\n",
    "                #        'Rhizosolenia': '#CC79A7',\n",
    "                #        'Thalassiosira': '#D55E00',\n",
    "                #        'Trieres': '#F8EF3A',\n",
    "                #        'Astrosyne': '#87cefa',\n",
    "                #        'Coscinodiscus': '#B5B5B5',\n",
    "                #        'Entomoneis': '#795548',\n",
    "                #        'Extubocellulus': '#607D8B',\n",
    "                #        'Gedaniella': '#3F51B5',\n",
    "                #        'Leptocylindrus': '#046E0A',\n",
    "                #        'Phaeodactylum': '#9E9E9E',\n",
    "                #        'Odontella': '#0072B2',\n",
    "                #        'Corethron': '#ffd700',\n",
    "                #        'Licmophora': '#98fb98',\n",
    "                #        'Proboscia': '#0000ff',\n",
    "                #        'Stephanopyxis': '#dda0dd',\n",
    "                #        'Synedropsis': '#673AB7',\n",
    "                #        'Aulacoseira': '#00BCD4',\n",
    "                #        'Skeletonema': '#fa8072',\n",
    "                #        'Eucampia': '#89CE00',\n",
    "                #        'Fragilariopsis': '#FFEB3B',\n",
    "                #        'Attheya': '#795548',\n",
    "                #        'Thalassionema': '#9C27B0',\n",
    "                #        'Craspedostauros': '#2196F3'\n",
    "                #    },\n",
    "                category_orders={aggregation_level: month_order,\n",
    "                                 \"genus\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                            'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                            'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', \n",
    "                                            'Entomoneis', 'Eucampia', 'Craspedostauros', 'Amphiprora', 'Guinardia', 'Triceratium']},\n",
    "                title=station)\n",
    "\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",\n",
    "            size=8,\n",
    "            color=\"#000000\"\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width=8.5 * pixels_per_cm,\n",
    "        height=7.5 * pixels_per_cm,\n",
    "        margin=dict(\n",
    "            l=0,\n",
    "            r=25,\n",
    "            b=25,\n",
    "            t=25\n",
    "        ),\n",
    "        xaxis_title_text='Month',\n",
    "        yaxis_title_text='Total diatom TPM',\n",
    "        legend_title_text=tax_level,\n",
    "        # Set range of x-axis\n",
    "        xaxis_range=[0, 300000]\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as png\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPM_per_month_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TPL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "#data_diatoms = data[(data['class'] == 'Bacillariophyta') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "data_diatoms = data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "\n",
    "# Remove transcripts below a certain TPL threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPL'] > 1]\n",
    "\n",
    "# Group by month, station and species, sum TPL\n",
    "data_diatoms = data_diatoms.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "# Specify the desired order of months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "               \"February_2021\", \"April_2021\", \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "## First, set conversion factor to transform pixels to cm\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "# Plot per station\n",
    "for station in data_diatoms[aggregation_level2].unique():\n",
    "    plot_data = data_diatoms[data_diatoms[aggregation_level2] == station]\n",
    "    # Normalise mean of TPL values to the total mean TPL of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPL / plot_data.groupby('month').TPL.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "    fig = px.histogram(plot_data, x = 'rel_expression_per_month', y=aggregation_level, color = tax_level,\n",
    "                 # Pick the colors for the stations\n",
    "                #color_discrete_map=(\n",
    "                #    {'Rare': '#808080', \n",
    "                #     'Asterionellopsis': '#ff8c00', \n",
    "                #     'Astrosyne': '#87cefa',\n",
    "                #     'Chaetoceros': '#E69F00',\n",
    "                #     'Corethron': '#ffd700',\n",
    "                #     'Coscinodiscus': '#B5B5B5',\n",
    "                #     'Craspedostauros': '#2196F3',\n",
    "                #     'Dactyliosolen': '#ba55d3',\n",
    "                #     'Ditylum': '#56B4E9',\n",
    "                #     'Entomoneis': '#795548',\n",
    "                #     'Eucampia': '#89CE00',\n",
    "                #     'Guinardia': '#009E73', \n",
    "                #     'Helicotheca': '#04686E',\n",
    "                #     'Leptocylindrus': '#046E0A',\n",
    "                #     'Odontella': '#0072B2',\n",
    "                #     'Pseudo-nitzschia': '#A42324',\n",
    "                #     'Rhizosolenia': '#CC79A7',\n",
    "                #     'Skeletonema': '#fa8072',\n",
    "                #     'Stephanopyxis': '#dda0dd',\n",
    "                #     'Thalassiosira': '#D55E00',\n",
    "                #     'Triceratium': '#CC7DFF',\n",
    "                #     'Amphiprora': '#c2ccc4' \n",
    "                #     }\n",
    "                #),\n",
    "                color_discrete_map={\n",
    "                        'Rare': '#808080',\n",
    "                        'Asterionellopsis': '#ff8c00',\n",
    "                        'Chaetoceros': '#E69F00',\n",
    "                        'Cyclotella': '#8BC34A',\n",
    "                        'Dactyliosolen': '#ba55d3',\n",
    "                        'Ditylum': '#56B4E9',\n",
    "                        'Helicotheca': '#04686E',\n",
    "                        'Pseudictyota': '#673AB7',\n",
    "                        'Pseudo-nitzschia': '#A42324',\n",
    "                        'Rhizosolenia': '#CC79A7',\n",
    "                        'Thalassiosira': '#D55E00',\n",
    "                        'Trieres': '#F8EF3A',\n",
    "                        'Astrosyne': '#87cefa',\n",
    "                        'Coscinodiscus': '#B5B5B5',\n",
    "                        'Entomoneis': '#795548',\n",
    "                        'Extubocellulus': '#607D8B',\n",
    "                        'Gedaniella': '#3F51B5',\n",
    "                        'Leptocylindrus': '#046E0A',\n",
    "                        'Phaeodactylum': '#9E9E9E',\n",
    "                        'Odontella': '#0072B2',\n",
    "                        'Corethron': '#ffd700',\n",
    "                        'Licmophora': '#98fb98',\n",
    "                        'Proboscia': '#0000ff',\n",
    "                        'Stephanopyxis': '#dda0dd',\n",
    "                        'Synedropsis': '#673AB7',\n",
    "                        'Aulacoseira': '#00BCD4',\n",
    "                        'Skeletonema': '#fa8072',\n",
    "                        'Eucampia': '#89CE00',\n",
    "                        'Fragilariopsis': '#FFEB3B',\n",
    "                        'Attheya': '#795548',\n",
    "                        'Thalassionema': '#9C27B0',\n",
    "                        'Craspedostauros': '#2196F3'\n",
    "                    },\n",
    "                category_orders={aggregation_level: month_order,\n",
    "                                 \"genus\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                            'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                            'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', \n",
    "                                            'Entomoneis', 'Eucampia', 'Craspedostauros', 'Amphiprora', 'Guinardia', 'Triceratium']},\n",
    "                title=station)\n",
    "\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",\n",
    "            size=8,\n",
    "            color=\"#000000\"\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width=8.5 * pixels_per_cm,\n",
    "        height=7.5 * pixels_per_cm,\n",
    "        margin=dict(\n",
    "            l=0,\n",
    "            r=25,\n",
    "            b=25,\n",
    "            t=25\n",
    "        ),\n",
    "        xaxis_title_text='Month',\n",
    "        yaxis_title_text='Total diatom TPL',\n",
    "        legend_title_text=tax_level,\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as png\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_TPL_per_month_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_TPL_per_month_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_TPL_per_month_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_relative_TPL_per_month_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "#data_diatoms = data[(data['class'] == 'Bacillariophyta') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "data_diatoms = data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "\n",
    "# Remove transcripts below a certain TPL threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPL'] > 1]\n",
    "\n",
    "# Group by month, station and species, sum TPL\n",
    "data_diatoms = data_diatoms.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "# Specify the desired order of months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "               \"February_2021\", \"April_2021\", \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "## First, set conversion factor to transform pixels to cm\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "# Plot per station\n",
    "for station in data_diatoms[aggregation_level2].unique():\n",
    "    plot_data = data_diatoms[data_diatoms[aggregation_level2] == station]\n",
    "    # Normalise mean of TPL values to the total mean TPL of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPL / plot_data.groupby('month').TPL.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "    fig = px.histogram(plot_data, x = 'TPL', y=aggregation_level, color = tax_level,\n",
    "                 # Pick the colors for the stations\n",
    "                #color_discrete_map=(\n",
    "                #    {'Rare': '#808080', \n",
    "                #     'Asterionellopsis': '#ff8c00', \n",
    "                #     'Astrosyne': '#87cefa',\n",
    "                #     'Chaetoceros': '#E69F00',\n",
    "                #     'Corethron': '#ffd700',\n",
    "                #     'Coscinodiscus': '#B5B5B5',\n",
    "                #     'Craspedostauros': '#2196F3',\n",
    "                #     'Dactyliosolen': '#ba55d3',\n",
    "                #     'Ditylum': '#56B4E9',\n",
    "                #     'Entomoneis': '#795548',\n",
    "                #     'Eucampia': '#89CE00',\n",
    "                #     'Guinardia': '#009E73', \n",
    "                #     'Helicotheca': '#04686E',\n",
    "                #     'Leptocylindrus': '#046E0A',\n",
    "                #     'Odontella': '#0072B2',\n",
    "                #     'Pseudo-nitzschia': '#A42324',\n",
    "                #     'Rhizosolenia': '#CC79A7',\n",
    "                #     'Skeletonema': '#fa8072',\n",
    "                #     'Stephanopyxis': '#dda0dd',\n",
    "                #     'Thalassiosira': '#D55E00',\n",
    "                #     'Triceratium': '#CC7DFF',\n",
    "                #     'Amphiprora': '#c2ccc4' \n",
    "                #     }\n",
    "                #),\n",
    "                color_discrete_map={\n",
    "                        'Rare': '#808080',\n",
    "                        'Asterionellopsis': '#ff8c00',\n",
    "                        'Chaetoceros': '#E69F00',\n",
    "                        'Cyclotella': '#8BC34A',\n",
    "                        'Dactyliosolen': '#ba55d3',\n",
    "                        'Ditylum': '#56B4E9',\n",
    "                        'Helicotheca': '#04686E',\n",
    "                        'Pseudictyota': '#673AB7',\n",
    "                        'Pseudo-nitzschia': '#A42324',\n",
    "                        'Rhizosolenia': '#CC79A7',\n",
    "                        'Thalassiosira': '#D55E00',\n",
    "                        'Trieres': '#F8EF3A',\n",
    "                        'Astrosyne': '#87cefa',\n",
    "                        'Coscinodiscus': '#B5B5B5',\n",
    "                        'Entomoneis': '#795548',\n",
    "                        'Extubocellulus': '#607D8B',\n",
    "                        'Gedaniella': '#3F51B5',\n",
    "                        'Leptocylindrus': '#046E0A',\n",
    "                        'Phaeodactylum': '#9E9E9E',\n",
    "                        'Odontella': '#0072B2',\n",
    "                        'Corethron': '#ffd700',\n",
    "                        'Licmophora': '#98fb98',\n",
    "                        'Proboscia': '#0000ff',\n",
    "                        'Stephanopyxis': '#dda0dd',\n",
    "                        'Synedropsis': '#673AB7',\n",
    "                        'Aulacoseira': '#00BCD4',\n",
    "                        'Skeletonema': '#fa8072',\n",
    "                        'Eucampia': '#89CE00',\n",
    "                        'Fragilariopsis': '#FFEB3B',\n",
    "                        'Attheya': '#795548',\n",
    "                        'Thalassionema': '#9C27B0',\n",
    "                        'Craspedostauros': '#2196F3'\n",
    "                    },\n",
    "                category_orders={aggregation_level: month_order,\n",
    "                                 \"genus\": ['Rare', 'Ditylum', 'Odontella', 'Thalassiosira', 'Chaetoceros', 'Rhizosolenia',\n",
    "                                            'Asterionellopsis', 'Coscinodiscus', 'Dactyliosolen', 'Helicotheca', 'Leptocylindrus',\n",
    "                                            'Proboscia', 'Pseudo-nitzschia', 'Corethron', 'Skeletonema', 'Stephanopyxis', 'Astrosyne', \n",
    "                                            'Entomoneis', 'Eucampia', 'Craspedostauros', 'Amphiprora', 'Guinardia', 'Triceratium']},\n",
    "                title=station)\n",
    "\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",\n",
    "            size=8,\n",
    "            color=\"#000000\"\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width=8.5 * pixels_per_cm,\n",
    "        height=7.5 * pixels_per_cm,\n",
    "        margin=dict(\n",
    "            l=0,\n",
    "            r=25,\n",
    "            b=25,\n",
    "            t=25\n",
    "        ),\n",
    "        xaxis_title_text='Month',\n",
    "        yaxis_title_text='Total diatom TPL',\n",
    "        legend_title_text=tax_level,\n",
    "        # Set range of x-axis\n",
    "        xaxis_range=[0, 650000000]\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as png\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPL_per_month_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPL_per_month_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPL_per_month_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatoms_total_TPL_per_month_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Biodiversity indices\n",
    "Now, we'll generate samples x species tables and use those to generate biodiversity indices per sample. Then, we'll visualize patterns in said indices over time (in months)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the data\n",
    "## Get the diatoms\n",
    "# tax_level = 'genus'\n",
    "tax_level = 'Name_to_Use'\n",
    "aggregation_level = 'sample'\n",
    "\n",
    "## Since we'll be looking at the relative abundance of different diatom genera, we can only include reads that are annotated to a genus with a sufficient % sequence identity\n",
    "data_diatoms = data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPM'] > 1]\n",
    "\n",
    "# First, I only want to include transcriptome bins that have at least 50 non-zero expressed transcripts in one of the samples\n",
    "# Group by month and genus, sum TPM\n",
    "data_diatoms = data_diatoms.groupby([aggregation_level, tax_level]).count().reset_index()\n",
    "# Now transform the data to the wide format\n",
    "data_diatoms = data_diatoms.pivot(index=tax_level, columns=aggregation_level, values='TPM')\n",
    "# Set NaN values to 0\n",
    "data_diatoms = data_diatoms.fillna(0)\n",
    "\n",
    "# Visualize the data\n",
    "data_diatoms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_diatoms[data_diatoms > 100].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many diatom genera exist per month that have > 100 non-zero TPM transcripts\n",
    "## Per sample, count the number of genera that have > 100 TPM\n",
    "taxonomic_bin_abundance = data_diatoms[data_diatoms > 100].count(axis=0).reset_index()\n",
    "taxonomic_bin_abundance.columns = ['sample', 'num_species']\n",
    "\n",
    "## Add metadata\n",
    "taxonomic_bin_abundance = taxonomic_bin_abundance.merge(meta[['month', 'station']], left_on='sample', right_index=True, how='left')\n",
    "\n",
    "# Plot the number of diatom genera per month that have > 100 TPM\n",
    "fig = px.box(taxonomic_bin_abundance, x='month', y='num_species',\n",
    "             category_orders={'month': ['July_2020', \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "               \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "               \"May_2021\", \"June_2021\", \"July_2021\"]})\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save the figure\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/num_eukprot_diatom_species_per_month.svg\", width = 3.5, height = 2, scale=1, format='svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ANOVA for months\n",
    "model_month = ols('num_species ~ month', data=taxonomic_bin_abundance).fit()\n",
    "anova_table_month = sm.stats.anova_lm(model_month, typ=2)\n",
    "\n",
    "# ANOVA for stations\n",
    "model_stations = ols('num_species ~ station', data=taxonomic_bin_abundance).fit()\n",
    "anova_table_stations = sm.stats.anova_lm(model_stations, typ=2)\n",
    "\n",
    "print(\"ANOVA Table for the difference in active species between months:\\n\", anova_table_month)\n",
    "print(\"\\nANOVA Table for the difference in active species between stations:\\n\", anova_table_stations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check normality of residuals for the months\n",
    "print(\"\\nChecking Normality of Residuals for Months...\")\n",
    "print(stats.shapiro(model_month.resid))\n",
    "sm.qqplot(model_month.resid, line='s')\n",
    "plt.title(\"Q-Q Plot of Residuals for Months\")\n",
    "plt.show()\n",
    "\n",
    "# Check homogeneity of variances for the months\n",
    "print(\"\\nChecking Homogeneity of Variances for the Months...\")\n",
    "print(stats.levene(*[taxonomic_bin_abundance['num_species'][taxonomic_bin_abundance['month'] == s] for s in taxonomic_bin_abundance['month'].unique()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Kruskal-Wallis test for the months as ANOVA assumptions were violated\n",
    "kw_table_month = stats.kruskal(*[taxonomic_bin_abundance['num_species'][taxonomic_bin_abundance['month'] == s] for s in taxonomic_bin_abundance['month'].unique()])\n",
    "\n",
    "print(\"\\nKruskal-Wallis Table for the difference in active species between months:\\n\", kw_table_month)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the genera that have > 100 TPM in at least one sample\n",
    "taxonomic_bin_abundance = data_diatoms[data_diatoms > 100].count(axis=1)\n",
    "taxonomic_bin_abundance.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot distribution of most abundant diatom genera\n",
    "taxonomic_bin_abundance.sort_values(ascending=False).plot(kind='bar', figsize=(10, 5));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the genera that have > 100 TPM in at least one sample\n",
    "taxonomic_bin_abundance = data_diatoms[data_diatoms > 100].count(axis=1).reset_index()\n",
    "taxonomic_bin_abundance.columns = ['taxonomic_bin', 'abundant_in_samples']\n",
    "taxonomic_bin_abundance.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Since we'll be looking at the relative abundance of different diatom genera, we can only include reads that are annotated to a genus with a sufficient % sequence identity\n",
    "data_diatoms = data[(data['Taxogroup2_UniEuk'] == 'Diatomeae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPM'] > 1]\n",
    "\n",
    "# Group by month and genus, sum TPM\n",
    "data_diatoms = data_diatoms.groupby([aggregation_level, tax_level]).count().reset_index()\n",
    "\n",
    "# Now transform the data to the wide format\n",
    "data_diatoms = data_diatoms.pivot(index=tax_level, columns=aggregation_level, values='TPM')\n",
    "# Set NaN values to 0\n",
    "data_diatoms = data_diatoms.fillna(0)\n",
    "\n",
    "# Filter the dataframe for these genera\n",
    "abundant_taxa = taxonomic_bin_abundance[taxonomic_bin_abundance['abundant_in_samples'] >= 1]['taxonomic_bin'].tolist()\n",
    "data_diatoms = data_diatoms[data_diatoms.index.isin(abundant_taxa)]\n",
    "\n",
    "# Save to csv\n",
    "data_diatoms.to_csv(f'../../data/analysis/diatom_{tax_level}_{aggregation_level}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now save the count of these genera per sample\n",
    "taxonomic_bin_abundance = data_diatoms[data_diatoms > 100].count(axis=0).reset_index()\n",
    "taxonomic_bin_abundance.columns = ['sample', 'num_species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomic_bin_abundance.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now run the above through the R [diversity](diversity.r) script. Then, we'll visualize the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load biodiversity estimates\n",
    "diatom_biodiversity = pd.read_csv('../../data/analysis/diatom_biodiversity_estimates.csv', index_col=0)\n",
    "\n",
    "# Add metadata to the biodiversity estimates\n",
    "meta = pd.read_csv('../../samples.csv', sep=';', index_col=0)\n",
    "diatom_biodiversity = diatom_biodiversity.merge(meta, left_index=True, right_index=True)\n",
    "\n",
    "# Add the number of diatom species per sample\n",
    "diatom_biodiversity = diatom_biodiversity.merge(taxonomic_bin_abundance, left_index=True, right_on='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot number of diatom genera per month\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(5*cm, 5*cm))\n",
    "\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxplot(x=\"month\", y=\"num_species\",\n",
    "              order=month_order,\n",
    "              data=diatom_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Change x axis labels rotation\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/diatom_species_per_month.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot evenness per month\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(5*cm, 5*cm))\n",
    "\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxenplot(x=\"month\", y=\"evenness\",\n",
    "              order=month_order,\n",
    "              scale=\"linear\", data=diatom_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Change x axis labels rotation\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "\n",
    "# Set y axis limits\n",
    "plt.ylim(-0.05, 1.05)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/diatom_evenness_per_month.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Calculate Functional Richness"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To the above dataframe, add the amount of unique KEGG KO ids per sample\n",
    "# Set aggregation levels of interest\n",
    "functional_info = 'KEGG_ko'\n",
    "tax_level = 'Name_to_Use'\n",
    "# tax_level = 'genus'\n",
    "aggregation_level = 'sample'\n",
    "# Only retain abundant species determined above\n",
    "data_diatoms = data[data[tax_level].isin(abundant_taxa)][[aggregation_level, functional_info, 'TPM']]\n",
    "\n",
    "# if processing kegg data, extra preprocessing is required: \n",
    "# Comment or uncomment the following line if multiple values are assigned to a transcript!\n",
    "## split them up\n",
    "data_diatoms = data_diatoms.assign(**{functional_info:data_diatoms[functional_info].str.split(',')})\n",
    "# Now we can explode the functional column\n",
    "data_diatoms = data_diatoms.explode(functional_info)\n",
    "# Cut off that weird part of the identifier, if necessary!\n",
    "data_diatoms[functional_info] = data_diatoms[functional_info].str.split(\":\", expand=True).drop(columns=0)\n",
    "\n",
    "# Remove the rows with no functional annotation\n",
    "data_diatoms = data_diatoms[data_diatoms[functional_info].notna()]\n",
    "data_diatoms = data_diatoms[data_diatoms[functional_info] != '-']\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_diatoms = data_diatoms[data_diatoms['TPM'] > 1]\n",
    "\n",
    "# Group by functional information and sample values, sum TPM\n",
    "functional_diversity = data_diatoms.groupby([functional_info, aggregation_level]).sum().reset_index()\n",
    "\n",
    "# Reshape the dataframe to wide format\n",
    "functional_diversity = functional_diversity.pivot(index=functional_info, columns=aggregation_level, values='TPM')\n",
    "\n",
    "# Set NaN values to 0\n",
    "functional_diversity = functional_diversity.fillna(0)\n",
    "\n",
    "# Calculate functional richness\n",
    "functional_diversity[functional_diversity < 1] = 0\n",
    "functional_diversity[functional_diversity >= 1] = 1\n",
    "\n",
    "functional_richness =  functional_diversity.sum(axis=0).reset_index()\n",
    "\n",
    "functional_richness.columns = ['sample', functional_info]\n",
    "functional_richness.set_index('sample', inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the functional richness to the dataframe\n",
    "diatom_biodiversity = diatom_biodiversity.merge(functional_richness, left_on='sample', right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diatom_biodiversity.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## This piece of code is now deprecated, but might still be useful in the future\n",
    "## Swap identifier for KO name\n",
    "## Load the KO names\n",
    "#KO_names = pd.read_csv('../../data/analysis/kegg_info.txt', sep='\\t', engine='pyarrow', header=None)\n",
    "#KO_names.columns = ['KO', 'Name']\n",
    "#\n",
    "## Remove the part in the name before the ;\n",
    "#KO_names['Name'] = KO_names['Name'].str.split(';', expand=True).drop(columns=[0,2])\n",
    "## Remove the part in the name between the []\n",
    "#KO_names['Name'] = KO_names['Name'].str.split('[', expand=True).drop(columns=[1,2,3])\n",
    "## Add the KO identifier to the name\n",
    "#KO_names['Name'] = KO_names['Name'] + ' (' + KO_names['KO'] + ')'\n",
    "#\n",
    "## Merge the KO names with the data\n",
    "#data_diatoms = data_diatoms.merge(KO_names, left_on=functional_info, right_on='KO', how='left')\n",
    "#\n",
    "## Remove the KO identifier\n",
    "#data_diatoms = data_diatoms.drop(columns=['KEGG_ko', 'KO'])\n",
    "#\n",
    "## Regroup to sum similar KO names\n",
    "#data_diatoms = data_diatoms.groupby(['Name', aggregation_level]).sum().reset_index()\n",
    "\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Abundance of functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the amount of KO per month in a boxplot\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(5*cm, 5*cm))\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxenplot(x=\"month\", y=\"KEGG_ko\",\n",
    "              order=month_order,\n",
    "              scale=\"linear\", data=diatom_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Change x axis labels rotation\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "\n",
    "# Set y axis limits\n",
    "plt.ylim(-50, 5500)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/diatom_KO_abundance_per_month.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Active species richnes vs abundance of functional modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There's fewer active species in dinoflagellates, so let's just create a boxplot of the number of active species vs functional richness\n",
    "# Plot the amount of KO per month in a boxplot\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(10*cm, 5*cm))\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxenplot(x=\"num_species\", y=\"KEGG_ko\",\n",
    "              scale=\"linear\", data=diatom_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Set y axis limits\n",
    "plt.ylim(-50, 5500)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/diatoms_KO_abundance_num_species.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How is functional richness correlated to the amount of diatom reads?\n",
    "## Calculate the number of reads that are assigned to diatoms in a given sample\n",
    "number_of_reads = data[data['Taxogroup2_UniEuk'] == 'Diatomeae'][[aggregation_level, 'TPM']]\n",
    "number_of_reads = number_of_reads.groupby(aggregation_level).sum().reset_index()\n",
    "\n",
    "# Add it to the diatom_biodiversity data\n",
    "diatom_biodiversity = diatom_biodiversity.merge(number_of_reads, on='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diatom_biodiversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "fig = px.scatter(diatom_biodiversity, x=\"TPM\", y=\"KEGG_ko\")\n",
    "\n",
    "fig.update_traces(marker=dict(size=5,\n",
    "                              color='#63C5DA',\n",
    "                              line=dict(width=1,\n",
    "                                        color='DarkSlateGrey')),\n",
    "                  selector=dict(mode='markers'))\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 5.5 * pixels_per_cm,\n",
    "    height= 5.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    plot_bgcolor='white',\n",
    "    showlegend=False\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey',\n",
    "    range = [-50, 3000]\n",
    ")\n",
    "\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_data = diatom_biodiversity.copy()\n",
    "\n",
    "# Define the models\n",
    "linear_model = LinearRegression()\n",
    "poly_model = make_pipeline(PolynomialFeatures(degree=2), LinearRegression())\n",
    "\n",
    "# Metrics to be calculated: 'neg_mean_squared_error' and 'r2'\n",
    "scoring_metrics = ['neg_mean_squared_error', 'r2']\n",
    "\n",
    "# Calculate the cross-validated scores for all models using mean squared error and r2 as the scoring metrics\n",
    "linear_scores = cross_validate(linear_model, plot_data[['num_species']], plot_data['KEGG_ko'], cv=5, scoring=scoring_metrics)\n",
    "poly_scores = cross_validate(poly_model, plot_data[['num_species']], plot_data['KEGG_ko'], cv=5, scoring=scoring_metrics)\n",
    "\n",
    "# Printing the results\n",
    "print(\"Linear model:\")\n",
    "print(f\"  Negative MSE: {linear_scores['test_neg_mean_squared_error'].mean():.2f} with a standard deviation of {linear_scores['test_neg_mean_squared_error'].std():.2f}\")\n",
    "print(f\"  R2: {linear_scores['test_r2'].mean():.4f} with a standard deviation of {linear_scores['test_r2'].std():.4f}\")\n",
    "\n",
    "print(\"\\nPolynomial model:\")\n",
    "print(f\"  Negative MSE: {poly_scores['test_neg_mean_squared_error'].mean():.2f} with a standard deviation of {poly_scores['test_neg_mean_squared_error'].std():.2f}\")\n",
    "print(f\"  R2: {poly_scores['test_r2'].mean():.4f} with a standard deviation of {poly_scores['test_r2'].std():.4f}\")\n",
    "\n",
    "# Performing a paired t-test to compare the performance of the two models on MSE\n",
    "t_stat, p_val = stats.ttest_rel(linear_scores['test_neg_mean_squared_error'], poly_scores['test_neg_mean_squared_error'])\n",
    "\n",
    "print(f\"\\nT-statistic for MSE: {t_stat}\")\n",
    "print(f\"P-value for MSE: {p_val}\")\n",
    "if p_val < 0.05:\n",
    "    print(\"The difference in MSE between the two models is statistically significant.\")\n",
    "else:\n",
    "    print(\"There is no statistically significant difference in MSE between the two models.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Linear model is the better fit\n",
    "## Fit the polynomial model\n",
    "linear_model.fit(plot_data[['num_species']], plot_data['KEGG_ko'])\n",
    "## Extract the predictions of the polynomial model\n",
    "# First generate x-values for the polynomial fit\n",
    "x_values = np.linspace(plot_data['num_species'].min(), plot_data['num_species'].max(), 100)\n",
    "\n",
    "# Predict y-values using the polynomial fit\n",
    "model_predictions = linear_model.predict(x_values.reshape(-1, 1))\n",
    "\n",
    "# Set conversion factor to transform pixels to cm\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "fig = px.scatter(plot_data, x=\"num_species\", y=\"KEGG_ko\")\n",
    "\n",
    "# Add the polynomial fit\n",
    "fig.add_trace(go.Scatter(\n",
    "    x=x_values,\n",
    "    y=model_predictions,\n",
    "    mode='lines',\n",
    "    line=dict(color='black', width=1),\n",
    "    name='Polynomial Fit'\n",
    "))\n",
    "\n",
    "fig.update_traces(marker=dict(size=5,\n",
    "                              color='#63C5DA',\n",
    "                              line=dict(width=1,\n",
    "                                        color='DarkSlateGrey')),\n",
    "                  selector=dict(mode='markers'))\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 5.5 * pixels_per_cm,\n",
    "    height= 5.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    plot_bgcolor='white',\n",
    "    showlegend=False\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey',\n",
    "    range = [-50, 3000]\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fit the data with scipy\n",
    "# Set the model, a negative exponential\n",
    "model = lambda x, a, b, c: b - (b-a) * np.exp(-c * x)\n",
    "popt, pcov = curve_fit(model, plot_data['num_species'], plot_data['KEGG_ko'])\n",
    "\n",
    "# Generate x-values for the exponential fit\n",
    "x_values = np.linspace(plot_data['num_species'].min(), plot_data['num_species'].max(), 100)\n",
    "# Predict y-values using the exponential fit\n",
    "model_predictions = model(x_values, *popt)\n",
    "\n",
    "fig = px.scatter(plot_data, x=\"num_species\", y=\"KEGG_ko\")\n",
    "\n",
    "# Add the polynomial fit\n",
    "fig.add_trace(go.Scatter(\n",
    "    x=x_values,\n",
    "    y=model_predictions,\n",
    "    mode='lines',\n",
    "    line=dict(color='black', width=1),\n",
    "    name='Polynomial Fit'\n",
    "))\n",
    "\n",
    "fig.update_traces(marker=dict(size=5,\n",
    "                              color='#63C5DA',\n",
    "                              line=dict(width=1,\n",
    "                                        color='DarkSlateGrey')),\n",
    "                  selector=dict(mode='markers'))\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 5.5 * pixels_per_cm,\n",
    "    height= 5.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    plot_bgcolor='white',\n",
    "    showlegend=False\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey',\n",
    "    range = [-50, 3000]\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save figure as png\n",
    "fig.write_image(f'../../figures/diatoms_vs_dinoflagellates/diatom_KO_abundance_vs_species.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Seems a better fit, let's compare both\n",
    "# Predict y-values using the linear model\n",
    "linear_predictions = linear_model.predict(plot_data[['num_species']])\n",
    "\n",
    "# Calculate MSE\n",
    "linear_mse = mean_squared_error(plot_data['KEGG_ko'], linear_predictions)\n",
    "\n",
    "# Calculate R-squared score\n",
    "linear_r2 = r2_score(plot_data['KEGG_ko'], linear_predictions)\n",
    "\n",
    "# Predict y-values using the exponential model\n",
    "exp_predictions = model(plot_data['num_species'], *popt)\n",
    "\n",
    "# Calculate MSE\n",
    "exp_mse = mean_squared_error(plot_data['KEGG_ko'], exp_predictions)\n",
    "\n",
    "# Calculate R-squared score\n",
    "exp_r2 = r2_score(plot_data['KEGG_ko'], exp_predictions)\n",
    "\n",
    "# Compare the two models\n",
    "print(f'Linear model: MSE = {linear_mse}, R2 = {linear_r2}')\n",
    "print(f'Exponential model: MSE = {exp_mse}, R2 = {exp_r2}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colour according to month\n",
    "fig = px.scatter(plot_data, x=\"num_species\", y=\"KEGG_ko\", trendline=\"ols\", color='month', color_discrete_sequence=px.colors.qualitative.Set1)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=14,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 14 * pixels_per_cm,\n",
    "    height= 10 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    plot_bgcolor='white',\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey',\n",
    "    range = [-50, 3000]\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save figure as png\n",
    "fig.write_image(f'../../figures/diatoms_vs_dinoflagellates/diatom_KO_abundance_vs_species_per_month.svg')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trophic feeding mode PFAMs\n",
    "In this part we'll use the list of PFAMs from the study by [Lambert & Groussman, 2022](https://www.pnas.org/doi/suppl/10.1073/pnas.2100916119#data-availability). In dataset S6, they provide a list of commmon PFAMs that capture functional diversity between different feeding modes. \n",
    "We'll mine our dataset for the expression of these PFAMs, and see if we can find any patterns. From the above we can hypothesize that the diatom community is photoautotrophic, and the dinoflagellate community is mixotrophic. We'll see if this is reflected in the expression of feeding-mode related PFAMS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the data from the paper, but skip first 3 rows\n",
    "relevant_pfams = pd.read_excel('../../data/external/pnas.2100916119.sd06.xlsx', skiprows=3)\n",
    "\n",
    "# Remove all NaN rows\n",
    "relevant_pfams = relevant_pfams.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "relevant_pfams.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(relevant_pfams['PFAM_ID'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, create a new data_diato dataframe with PFAM expression data\n",
    "functional_info = 'PFAMs'\n",
    "aggregation_level = 'month'\n",
    "data_diatoms_PFAM = data[data['class'] == 'Bacillariophyta'][[aggregation_level, functional_info, 'TPM']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many of the relevant PFAMs are present in the data\n",
    "data_diatoms_PFAM[functional_info].isin(relevant_pfams['short_name']).value_counts(normalize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check to which clade these hits belong\n",
    "relevant_pfams[relevant_pfams['short_name'].isin(data_diatoms_PFAM[functional_info])]['clade'].value_counts()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looks promising, most PFAMs occur in the class that is linked to photoautotrophs. Let's check which of the relevant PFAMs is most expressed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only retain the relevant PFAMs\n",
    "data_diatoms_PFAM = data_diatoms_PFAM[data_diatoms_PFAM[functional_info].isin(relevant_pfams['short_name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What are the most abundant PFAMs?\n",
    "data_diatoms_PFAM.groupby(functional_info)['TPM'].sum().sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_diatoms_PFAM.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Determine the genes with highest variance across samples\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = data_diatoms_PFAM.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Add number of stations visited per month\n",
    "samples_per_month = {\n",
    "    'July_2020': 6,\n",
    "    'August_2020': 6,\n",
    "    'September_2020': 6,\n",
    "    'November_2020': 6,\n",
    "    'December_2020': 6,\n",
    "    'January_2021': 5,\n",
    "    'February_2021': 5,\n",
    "    'April_2021': 4,\n",
    "    'May_2021': 6,\n",
    "    'June_2021': 6,\n",
    "    'July_2021': 6\n",
    "}\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to diatoms\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "# Re-order the columns according to month\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# In the index, include the PFAM clade and functional descriptions to the short names between brackets\n",
    "plot_data_scaled.index = plot_data_scaled.index.map(lambda x: \"{} ({} - {})\".format(x, relevant_pfams[relevant_pfams['short_name'] == x]['clade'].values[0], relevant_pfams[relevant_pfams['short_name'] == x]['function'].values[0]))\n",
    "\n",
    "# Only retain the 15 most variable PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.var(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False,\n",
    "    # Legend bar should be on the right\n",
    "    cbar_kws={'orientation': 'vertical'},\n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Save figure as svg\n",
    "g.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatoms_heatmap_{}_most_variable_trophic_{}_per_{}.svg\".format(number_of_genes, functional_info, aggregation_level), format='svg', dpi=600)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'll do the same but with the 15 most expressed PFAMs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Determine the genes with highest variance across samples\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = data_diatoms_PFAM.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Add number of stations visited per month\n",
    "samples_per_month = {\n",
    "    'July_2020': 6,\n",
    "    'August_2020': 6,\n",
    "    'September_2020': 6,\n",
    "    'November_2020': 6,\n",
    "    'December_2020': 6,\n",
    "    'January_2021': 5,\n",
    "    'February_2021': 5,\n",
    "    'April_2021': 4,\n",
    "    'May_2021': 6,\n",
    "    'June_2021': 6,\n",
    "    'July_2021': 6\n",
    "}\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to diatoms\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "# Re-order the columns according to month\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# In the index, include the PFAM clade and functional descriptions to the short names between brackets\n",
    "plot_data_scaled.index = plot_data_scaled.index.map(lambda x: \"{} ({} - {})\".format(x, relevant_pfams[relevant_pfams['short_name'] == x]['clade'].values[0], relevant_pfams[relevant_pfams['short_name'] == x]['function'].values[0]))\n",
    "\n",
    "# Only retain the 15 most expressed PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.sum(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]\n",
    "\n",
    "# \n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False,\n",
    "    # Legend bar should be on the right\n",
    "    cbar_kws={'orientation': 'vertical'},\n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Save figure as svg\n",
    "g.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatoms_heatmap_{}_most_expressed_trophic_{}_per_{}.svg\".format(number_of_genes, functional_info, aggregation_level), format='svg', dpi=600)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we'll do the same with the full feature set (all 1046 gene families used in the model described in the paper)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PFAM_features = pd.read_excel('../../data/external/pnas.2100916119.sd03.xlsx', skiprows=1, header=1)\n",
    "\n",
    "PFAM_features = PFAM_features.dropna()\n",
    "\n",
    "PFAM_features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Of the PFAM features, which are most expressed?\n",
    "## Subset the data to only include the PFAMs\n",
    "functional_info = 'PFAMs'\n",
    "aggregation_level = 'month'\n",
    "data_diatoms_PFAM = data[data['class'] == 'Bacillariophyta'][[aggregation_level, functional_info, 'TPM']]\n",
    "diatom_features = data_diatoms_PFAM[data_diatoms_PFAM[functional_info].isin(PFAM_features['Name'])]\n",
    "\n",
    "# Group by month, PFAM and sum the TPM values\n",
    "diatom_features = diatom_features.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# What are the most abundant PFAMs?\n",
    "diatom_features.groupby(functional_info)['TPM'].sum().sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the most abundant PFAMs\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = diatom_features.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to diatoms\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# Only retain the 15 most variable PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.var(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False,\n",
    "    # Legend bar should be on the right\n",
    "    cbar_kws={'orientation': 'vertical'},\n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Save figure as svg\n",
    "# g.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatoms_heatmap_{}_most_variable_trophic_{}_per_{}.svg\".format(number_of_genes, functional_info, aggregation_level), format='svg', dpi=600)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we'll do the same but with the 15 most highly expressed PFAMs\n",
    "# Visualize the most abundant PFAMs\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = diatom_features.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to diatoms\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# Only retain the 15 most expressed PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.sum(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]\n",
    "\n",
    "# \n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False,\n",
    "    # Legend bar should be on the right\n",
    "    cbar_kws={'orientation': 'vertical'},\n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Save figure as svg\n",
    "# g.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatoms_heatmap_{}_most_variable_trophic_{}_per_{}.svg\".format(number_of_genes, functional_info, aggregation_level), format='svg', dpi=600)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trophic feeding mode of diatoms\n",
    "I've predicted the trophic feeding mode of several species using the machine learning model of the paper by [Lambert & Groussman, 2022](https://doi.org/10.1073/pnas.2100916119), see [here](./trophic_mode_prediction.ipynb). We'll use the data to see how the trophic feeding mode of diatoms changes over time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "#trophic_predictions = pd.read_csv('../../data/analysis/phylodb_trophic_mode_predictions.csv')\n",
    "trophic_predictions = pd.read_csv('../../data/analysis/eukprot_trophic_mode_predictions.csv')\n",
    "# Only retain the diatoms\n",
    "#trophic_predictions = trophic_predictions[trophic_predictions['class'] == 'Bacillariophyta']\n",
    "trophic_predictions = trophic_predictions[trophic_predictions['Taxogroup2_UniEuk'] == 'Diatomeae']\n",
    "print(len(trophic_predictions))\n",
    "trophic_predictions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the relative abundance of the three trophic modes per month\n",
    "# Calculate relative abundance of each prediction per month\n",
    "rel_abundance = (trophic_predictions[['month', 'station', 'prediction']].groupby(['month', 'prediction'])\n",
    "                       .size()\n",
    "                       .groupby(level=0).apply(lambda x: 100 * x / x.sum())\n",
    "                       .reset_index(name='relative abundance'))\n",
    "# Set the color palette\n",
    "color_map = {\n",
    "    'Phot': 'green',\n",
    "    'Mix': 'black',\n",
    "    'Het': 'red'\n",
    "}\n",
    "fig = px.histogram(rel_abundance.sort_values(\"month\", ascending=False),\n",
    "             x = \"relative abundance\",\n",
    "             y = \"month\",\n",
    "             color = \"prediction\",\n",
    "             orientation='h',\n",
    "             color_discrete_map=color_map,\n",
    "             category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"]})\n",
    "\n",
    "pixels_per_cm = 37.7952755906\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8  # Set the font size\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Ratio of predicted trophic modes (%)',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatom_trophic_mode_relative_abundance_phylodb.svg\")\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/diatom_trophic_mode_relative_abundance_eukprot.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the relevant columns\n",
    "# tax_level='genus'\n",
    "tax_level='Genus_UniEuk'\n",
    "\n",
    "df = trophic_predictions[[tax_level, 'month', 'prediction']]\n",
    "# Convert 'prediction' into numerical categories\n",
    "df['prediction'] = df['prediction'].map({'Phot':1, 'Mix':2, 'Het':3})\n",
    "\n",
    "# Pivot the DataFrame to use for heatmap\n",
    "df_pivot = df.pivot_table(index=tax_level, columns='month', values='prediction', \n",
    "                          aggfunc=lambda x: int(stats.mode(x[x!=0])[0][0] if any(x!=0) else 0))\n",
    "\n",
    "# Fill NA with a specific category (e.g., 0 for 'no prediction')\n",
    "df_pivot = df_pivot.fillna(0)\n",
    "\n",
    "# Ensure the values are integer\n",
    "df_pivot = df_pivot.astype(int)\n",
    "\n",
    "# Set the order of the months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "df_pivot = df_pivot.reindex(month_order, axis=1)\n",
    "\n",
    "# Set the order of the genera to be the same as in the above relative abundance plot\n",
    "df_pivot = df_pivot.reindex(diatom_legend_order, axis=0)\n",
    "\n",
    "# Create a color map\n",
    "cmap = mcolors.ListedColormap(['lightgrey', 'green', 'yellow', 'red'])\n",
    "\n",
    "# Create the heatmap with the color map\n",
    "plt.figure(figsize=(4.5, 4))\n",
    "sns.heatmap(df_pivot, cmap=cmap, annot=False, cbar=False, vmin=0, vmax=3)\n",
    "\n",
    "# Change layout to match the relative abundance plot\n",
    "plt.tight_layout()\n",
    "\n",
    "# Remove the x and  y-axis label\n",
    "plt.xlabel(None)\n",
    "plt.ylabel(None)\n",
    "\n",
    "# Font size should be 8, and font family Times New Roman\n",
    "plt.rcParams.update({'font.size': 8, 'font.family': 'Times New Roman'})\n",
    "\n",
    "# Save figure as svg\n",
    "#plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatom_trophic_mode_genus_consensus_heatmap_phylodb.svg\", format='svg', dpi=600)\n",
    "plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatom_trophic_mode_genus_consensus_heatmap_eukprot.svg\", format='svg', dpi=600)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PCA"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For a PCA, we need a new matrix samples x genes. We'll also need a new metadata dataframe with the samples, months, stations, and environmental variables. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify how the matrix will look like (in the end its aggregation_level x functional_info)\n",
    "functional_info = 'KEGG_ko'\n",
    "aggregation_level = 'sample'\n",
    "\n",
    "# Prepare gene expression data and load environmental variables\n",
    "#data_diatoms = data[data['class'] == 'Bacillariophyta'][[aggregation_level, functional_info, 'TPM']]\n",
    "data_diatoms = data[data['Taxogroup2_UniEuk'] == 'Diatomeae'][[aggregation_level, functional_info, 'TPM']]\n",
    "\n",
    "# if processing kegg data, extra preprocessing is required: \n",
    "# Comment or uncomment the following line if multiple values are assigned to a transcript!\n",
    "## split KEGG identifiers up\n",
    "data_diatoms = data_diatoms.assign(**{functional_info:data_diatoms[functional_info].str.split(',')})\n",
    "# Now we can explode the functional column\n",
    "data_diatoms = data_diatoms.explode(functional_info)\n",
    "# Cut off that weird part of the identifier, if necessary!\n",
    "data_diatoms[functional_info] = data_diatoms[functional_info].str.split(\":\", expand=True).drop(columns=0)\n",
    "\n",
    "# Group by functional information and sample values, sum TPM\n",
    "data_diatoms = data_diatoms.groupby([functional_info, aggregation_level]).sum().reset_index()\n",
    "\n",
    "# Remove the rows with no functional annotation\n",
    "data_diatoms = data_diatoms[data_diatoms[functional_info].notna()]\n",
    "data_diatoms = data_diatoms[data_diatoms[functional_info] != '-']\n",
    "\n",
    "# Transform the data to the wide format\n",
    "data_diatoms = data_diatoms.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "\n",
    "# Log2 transform the data\n",
    "data_diatoms_log = np.log2(data_diatoms + 1)\n",
    "\n",
    "# Scale the features (columns, TPM values of every prediction)\n",
    "data_diatoms_scaled = pd.DataFrame(StandardScaler().fit_transform(data_diatoms_log),\n",
    "                                   index=data_diatoms.index,\n",
    "                                   columns=data_diatoms.columns)\n",
    "## Scaling removes the mean and scales to unit variance, the resulting values are z-scores\n",
    "\n",
    "# Load the environmental variables\n",
    "env_variables = pd.read_csv(\"../../data/environmental/samples_env.csv\", sep=\";\")\n",
    "env_variables = env_variables.set_index('sample')\n",
    "\n",
    "# Merge environmental variables with gene expression data\n",
    "pca_data = data_diatoms_log.merge(env_variables, left_index=True, right_on='sample', how='left')\n",
    "\n",
    "# Perform PCA on gene expression data, ignore columns that are also in the environmental variables\n",
    "pca = PCA(n_components=2)\n",
    "principalComponents = pca.fit_transform(pca_data.drop(columns=env_variables.columns))\n",
    "\n",
    "# Add principal components to the data\n",
    "pca_data['PC1'] = principalComponents[:, 0]\n",
    "pca_data['PC2'] = principalComponents[:, 1]\n",
    "\n",
    "# Create a color dictionary for the months\n",
    "month_color_dict = dict(zip(data['month'].unique(), sns.color_palette('tab10', n_colors=len(data['month'].unique()))))\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Set figure size and font scale\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(12*cm, 14*cm))\n",
    "sns.set(style='white', font_scale=1)\n",
    "\n",
    "# Plot PCA biplot with the color of the month corrresponding to the month of the sample\n",
    "# and the shape of the point corresponding to the station of the sample\n",
    "sns.scatterplot(data=pca_data,\n",
    "                x='PC1',\n",
    "                y='PC2',\n",
    "                hue='month',\n",
    "                hue_order=month_order,\n",
    "                style='station',\n",
    "                palette=month_color_dict,\n",
    "                s=40, edgecolor='black', linewidth=0.5)\n",
    "\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0)\n",
    "# Plot labels with the variance explained by each principal component\n",
    "plt.xlabel('Principal Component 1 ({}%)'.format(round(pca.explained_variance_ratio_[0] * 100, 2)))\n",
    "plt.ylabel('Principal Component 2 ({}%)'.format(round(pca.explained_variance_ratio_[1] * 100, 2)))\n",
    "\n",
    "# Plot arrows indicating the correlation between the principal components and the environmental variables\n",
    "## Linear regression to determine the correlation between the principal components and the environmental variables\n",
    "columns_of_interest = ['NO3', 'PO4', 'Si', 'SPM', 'salinity', 'Temperature']\n",
    "\n",
    "# Calculate the correlation matrix between the principal components and the environmental variables\n",
    "corr_matrix = np.corrcoef(pca_data[['PC1', 'PC2'] + columns_of_interest].dropna().T)\n",
    "\n",
    "# Get the correlation between the first two principal components and each environmental parameter\n",
    "corr_PC1_env = corr_matrix[0, 2:]\n",
    "corr_PC2_env = corr_matrix[1, 2:]\n",
    "\n",
    "# Plot the arrows representing the correlations\n",
    "for i, env_param in enumerate(columns_of_interest):\n",
    "    plt.arrow(0, 0, corr_PC1_env[i]*max(pca_data['PC1']), corr_PC2_env[i]*max(pca_data['PC2']), head_width=0.05, color='gray')\n",
    "    # corr_PC1_env[i] specifies the end coordinate of the arrow in the x direction\n",
    "    # This coordinate is multiplied with max(pca_data['PC1']) to scale the arrow to the length of the principal component for visualization\n",
    "    plt.text(corr_PC1_env[i]*max(pca_data['PC1'])*1.05, corr_PC2_env[i]*max(pca_data['PC2'])*1.05, env_param, fontsize=12, color='gray')\n",
    "    # Same as above, but the text is placed 5% further away from the arrow\n",
    "\n",
    "# Save figure as svg\n",
    "# plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatom_PCA_{}_per_{}_phylodb.svg\".format(functional_info, aggregation_level), format='svg', dpi=600)\n",
    "plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/diatom_PCA_{}_per_{}_eukprot.svg\".format(functional_info, aggregation_level), format='svg', dpi=600)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save diatom expression data to a csv file\n",
    "#data_diatoms.to_csv('../../data/analysis/kegg_expression_diatoms_phylodb.csv')\n",
    "data_diatoms.to_csv('../../data/analysis/kegg_expression_diatoms_eukprot.csv')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Dinoflagellates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, let's get a list of all the unique dinoflagellate species\n",
    "#data[(data['class'] == 'Dinophyceae')  & (data['p_ident'] >= 0.98)]['species'].unique()\n",
    "data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales']))  & (data['p_ident'] >= 0.98)]['Name_to_Use'].unique()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Dinoflagellate abundance\n",
    "For dinoflagellates, we'll do the same analysis as for diatoms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "\n",
    "#data_dinoflagellates = data[(data['class'] == 'Dinophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# Group by month and species, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, tax_level]).sum().reset_index()\n",
    "\n",
    "data_dinoflagellates['month'] = pd.Categorical(data_dinoflagellates['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Normalise mean of TPM values to the total mean TPM of that month\n",
    "data_dinoflagellates[\"rel_expression_per_month\"] = data_dinoflagellates.TPM / data_dinoflagellates.groupby('month').TPM.transform('sum')\n",
    "\n",
    "# Combine low-abundant groups\n",
    "rare_groups = data_dinoflagellates['rel_expression_per_month'] <= 0.02\n",
    "data_dinoflagellates.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "# Print unique species\n",
    "print(data_dinoflagellates[tax_level].unique())\n",
    "# Inspect data\n",
    "data_dinoflagellates.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add conversion variable\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "# Plot\n",
    "fig = px.histogram(data_dinoflagellates.sort_values(\"month\", ascending=False), \n",
    "                x=\"rel_expression_per_month\", \n",
    "                y=\"month\", \n",
    "                color=tax_level,\n",
    "                orientation='h',\n",
    "                color_discrete_map={\n",
    "                    \"Rare\": \"#545454\",\n",
    "                    \"Alexandrium\": \"#ebac23\",\n",
    "                    \"Amphidinium\": \"#b80058\",\n",
    "                    \"Noctiluca\": \"#008cf9\",\n",
    "                    \"Tripos\": \"#006e00\",\n",
    "                    \"Kryptoperidinium\": \"#00bbad\",\n",
    "                    \"Oxyrrhis\": \"#d163e6\",\n",
    "                    \"Karenia\": \"#b24502\",\n",
    "                    \"Symbiodinium\": \"#ff9287\",\n",
    "                    \"Ceratium\": \"#5954d6\",\n",
    "                    \"Durinskia\": \"#00c6f8\",\n",
    "                    \"Heterocapsa\": \"#878500\",\n",
    "                    \"Scripsiella\": \"#00a76c\",\n",
    "                    'Polarella': \"#FC8D62\",\n",
    "                    'Gonyaulax': \"#8DA0CB\",\n",
    "                    'Azadinium': \"#E78AC3\",\n",
    "                    'Lingulodinium': \"#A6D854\",\n",
    "                    'Karlodinium': \"#B3B3CC\"\n",
    "                },\n",
    "                category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                                            \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                                            \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                                            \"June_2021\", \"July_2021\"],\n",
    "                \"Genus_UniEuk\": [\"Rare\", \"Tripos\", \"Noctiluca\"]},\n",
    "                # text_auto='.2f'\n",
    "                )\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='% TPM of total sum',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save figure as png\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_phylodb.png\".format(tax_level), scale=1)\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_phylodb.svg\".format(tax_level), scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_eukprot.png\".format(tax_level), scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_eukprot.svg\".format(tax_level), scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store the order of the genera in the legend\n",
    "dinoflagellate_legend_order = []\n",
    "for i in range(len(fig.data)):\n",
    "    dinoflagellate_legend_order.append(fig.data[i].name)\n",
    "print(dinoflagellate_legend_order)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Spatial distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "#data_dinoflagellates = data[(data['class'] == 'Dinophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# Group by month, station and species, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "data_dinoflagellates['month'] = pd.Categorical(data_dinoflagellates['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "dinoflagellate_genera = []\n",
    "# Plot per station\n",
    "for station in data_dinoflagellates[aggregation_level2].unique():\n",
    "    plot_data = data_dinoflagellates[data_dinoflagellates[aggregation_level2] == station]\n",
    "    # Normalise mean of TPM values to the total mean TPM of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPM / plot_data.groupby('month').TPM.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "    # Add all unique genera to dictionary if not already present\n",
    "    for genus in plot_data[tax_level].unique():\n",
    "        dinoflagellate_genera.append(genus) if genus not in dinoflagellate_genera else None\n",
    "    fig = px.histogram(plot_data.sort_values(\"month\", ascending=False), \n",
    "                    x=\"rel_expression_per_month\", \n",
    "                    y=\"month\", \n",
    "                    color=tax_level,\n",
    "                    orientation='h',\n",
    "                    # text_auto='.2f',\n",
    "                    color_discrete_map={\n",
    "                        \"Rare\": \"#545454\",\n",
    "                        \"Alexandrium\": \"#ebac23\",\n",
    "                        \"Amphidinium\": \"#b80058\",\n",
    "                        \"Noctiluca\": \"#008cf9\",\n",
    "                        \"Tripos\": \"#006e00\",\n",
    "                        \"Kryptoperidinium\": \"#00bbad\",\n",
    "                        \"Oxyrrhis\": \"#d163e6\",\n",
    "                        \"Karenia\": \"#b24502\",\n",
    "                        \"Symbiodinium\": \"#ff9287\",\n",
    "                        \"Ceratium\": \"#5954d6\",\n",
    "                        \"Durinskia\": \"#00c6f8\",\n",
    "                        \"Heterocapsa\": \"#878500\",\n",
    "                        \"Scripsiella\": \"#00a76c\",\n",
    "                        'Polarella': \"#FC8D62\",\n",
    "                        'Gonyaulax': \"#8DA0CB\",\n",
    "                        'Azadinium': \"#E78AC3\",\n",
    "                        'Lingulodinium': \"#A6D854\",\n",
    "                        'Karlodinium': \"#B3B3CC\"\n",
    "                    },\n",
    "                    # Specify all the months that need to be included, \n",
    "                    # even if no sample has been taken\n",
    "                    category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                                        \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                                        \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                                        \"June_2021\", \"July_2021\"],\n",
    "                                     \"genus\": [\"Rare\", \"Alexandrium\", \"Amphidinium\", \"Noctiluca\", \n",
    "                                               \"Symbiodinium\", \"Ceratium\", \"Tripos\", \"Kryptoperidinium\", \n",
    "                                               \"Oxyrrhis\", \"Karenia\", \"Durinskia\", \"Heterocapsa\", \"Scripsiella\"],\n",
    "                                     #\"Genus_UniEuk\": [\"Rare\", \"Noctiluca\", \"Tripos\", \"Paragymnodinium\", \"Pelagodinium\", \"Polykrikos\"]\n",
    "                                     }\n",
    "                    )\n",
    "\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "            size=8,  # Set the font size\n",
    "            color=\"#000000\"  # Set the font color\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width= 8.5 * pixels_per_cm,\n",
    "        height= 7.5 * pixels_per_cm,\n",
    "        margin=dict( # Set the margins\n",
    "            l=0,  # Left margin\n",
    "            r=25,  # Right margin\n",
    "            b=25,  # Bottom margin\n",
    "            t=25  # Top margin\n",
    "        ),\n",
    "        xaxis_title_text='% TPM of total sum',\n",
    "        yaxis_title_text=None,\n",
    "        title_text=station\n",
    "    )\n",
    "\n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as png\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_expression_per_month_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dinoflagellate_genera\n",
    "# Now create colours for all of these genera and rerun the above code"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Total TPM of dinoflagellates per month"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "\n",
    "#data_dinoflagellates = data[(data['class'] == 'Dinophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# Group by month and species, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, tax_level]).sum().reset_index()\n",
    "\n",
    "data_dinoflagellates['month'] = pd.Categorical(data_dinoflagellates['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Normalise mean of TPM values to the total mean TPM of that month\n",
    "data_dinoflagellates[\"rel_expression_per_month\"] = data_dinoflagellates.TPM / data_dinoflagellates.groupby('month').TPM.transform('sum')\n",
    "\n",
    "# Combine low-abundant groups\n",
    "rare_groups = data_dinoflagellates['rel_expression_per_month'] <= 0.02\n",
    "data_dinoflagellates.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "# Plot the total TPM per month\n",
    "pixels_per_cm = 37.79527559055118\n",
    "fig = px.histogram(data_dinoflagellates.sort_values(\"month\", ascending=False),\n",
    "            x = \"TPM\",\n",
    "            y = \"month\",\n",
    "            color = tax_level,\n",
    "            color_discrete_map={\n",
    "                \"Rare\": \"#545454\",\n",
    "                \"Alexandrium\": \"#ebac23\",\n",
    "                \"Amphidinium\": \"#b80058\",\n",
    "                \"Noctiluca\": \"#008cf9\",\n",
    "                \"Tripos\": \"#006e00\",\n",
    "                \"Kryptoperidinium\": \"#00bbad\",\n",
    "                \"Oxyrrhis\": \"#d163e6\",\n",
    "                \"Karenia\": \"#b24502\",\n",
    "                \"Symbiodinium\": \"#ff9287\",\n",
    "                \"Ceratium\": \"#5954d6\",\n",
    "                \"Durinskia\": \"#00c6f8\",\n",
    "                \"Heterocapsa\": \"#878500\",\n",
    "                \"Scripsiella\": \"#00a76c\",\n",
    "                'Polarella': \"#FC8D62\",\n",
    "                'Gonyaulax': \"#8DA0CB\",\n",
    "                'Azadinium': \"#E78AC3\",\n",
    "                'Lingulodinium': \"#A6D854\",\n",
    "                'Karlodinium': \"#B3B3CC\"\n",
    "            },\n",
    "            #color_discrete_map={\"Rare\": \"#545454\", \"Noctiluca\": \"#008cf9\", \"Tripos\": \"#006e00\",\n",
    "            #                         \"Paragymnodinium\": \"#9932CC\", \"Pelagodinium\": \"#40E0D0\",\n",
    "            #                         \"Polykrikos\": \"#FFA500\"},\n",
    "            orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"],\n",
    "                \"genus\": dinoflagellate_legend_order}\n",
    "                #\"Genus_UniEuk\": [\"Rare\", \"Tripos\", \"Noctiluca\"]},\n",
    "            )\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total dinoflagellate TPM per month',\n",
    "    yaxis_title_text=None,\n",
    "    # Set range of x-axis\n",
    "    xaxis_range=[0, 550000],\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_per_month_{tax_level}_phylodb.svg\", scale=1)\n",
    "fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_per_month_{tax_level}_eukprot.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the total TPM per month per station\n",
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "#data_dinoflagellates = data[(data['class'] == 'Dinophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# Group by month and species, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "data_dinoflagellates['month'] = pd.Categorical(data_dinoflagellates['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Normalise mean of TPM values to the total mean TPM of that month\n",
    "data_dinoflagellates[\"rel_expression_per_month\"] = data_dinoflagellates.TPM / data_dinoflagellates.groupby('month').TPM.transform('sum')\n",
    "\n",
    "# Combine low-abundant groups\n",
    "rare_groups = data_dinoflagellates['rel_expression_per_month'] <= 0.02\n",
    "data_dinoflagellates.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "fig = px.histogram(data_dinoflagellates.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = 'station',\n",
    "            color_discrete_map={\n",
    "                \"ZG02\": \"#8c613c\",\n",
    "                \"120\": \"#956cb4\",\n",
    "                \"330\": \"#ee854a\",\n",
    "                \"130\": \"#4878d0\",\n",
    "                \"780\": \"#d65f5f\",\n",
    "                \"700\": \"#6acc64\"},\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"]},)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total dinoflagellate TPM per month',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_per_month_station_phylodb.svg\", scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_per_month_station_eukprot.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Spatial distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the taxonomic level and aggregation levels\n",
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "# Filter the data for dinoflagellates\n",
    "#data_dinoflagellates = data[(data['class'] == 'Dinophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "#data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# Group by month, station, and species, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "# Specify the desired order of months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "               \"February_2021\", \"April_2021\", \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Plot total TPM per station\n",
    "pixels_per_cm = 37.79527559055118\n",
    "for station in data_dinoflagellates[aggregation_level2].unique():\n",
    "    plot_data = data_dinoflagellates[data_dinoflagellates[aggregation_level2] == station]\n",
    "    # Normalise mean of TPM values to the total mean TPM of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPM / plot_data.groupby('month').TPM.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "    \n",
    "    fig = px.histogram(plot_data, x='TPM', y=aggregation_level, color=tax_level,\n",
    "            color_discrete_map={\n",
    "                \"Rare\": \"#545454\",\n",
    "                \"Alexandrium\": \"#ebac23\",\n",
    "                \"Amphidinium\": \"#b80058\",\n",
    "                \"Noctiluca\": \"#008cf9\",\n",
    "                \"Tripos\": \"#006e00\",\n",
    "                \"Kryptoperidinium\": \"#00bbad\",\n",
    "                \"Oxyrrhis\": \"#d163e6\",\n",
    "                \"Karenia\": \"#b24502\",\n",
    "                \"Symbiodinium\": \"#ff9287\",\n",
    "                \"Ceratium\": \"#5954d6\",\n",
    "                \"Durinskia\": \"#00c6f8\",\n",
    "                \"Heterocapsa\": \"#878500\",\n",
    "                \"Scripsiella\": \"#00a76c\",\n",
    "                'Polarella': \"#FC8D62\",\n",
    "                'Gonyaulax': \"#8DA0CB\",\n",
    "                'Azadinium': \"#E78AC3\",\n",
    "                'Lingulodinium': \"#A6D854\",\n",
    "                'Karlodinium': \"#B3B3CC\"\n",
    "            },\n",
    "            category_orders={aggregation_level: month_order},\n",
    "            title=station)\n",
    "    \n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",\n",
    "            size=8,\n",
    "            color=\"#000000\"\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width=8.5 * pixels_per_cm,\n",
    "        height=7.5 * pixels_per_cm,\n",
    "        margin=dict(\n",
    "            l=0,\n",
    "            r=25,\n",
    "            b=25,\n",
    "            t=25\n",
    "        ),\n",
    "        xaxis_title_text='Month',\n",
    "        yaxis_title_text='Total dinoflagellate TPM',\n",
    "        legend_title_text=tax_level,\n",
    "        # Set range of x-axis\n",
    "        xaxis_range=[0, 250000]\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as PNG and SVG\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPM_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TPL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the taxonomic level and aggregation levels\n",
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "# Filter the data for dinoflagellates\n",
    "#data_dinoflagellates = data[(data['class'] == 'Dinophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "\n",
    "# Remove transcripts below a certain TPL threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPL'] > 1]\n",
    "\n",
    "# Group by month, station, and species, sum TPL\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "# Specify the desired order of months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "               \"February_2021\", \"April_2021\", \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Plot total TPL per station\n",
    "pixels_per_cm = 37.79527559055118\n",
    "for station in data_dinoflagellates[aggregation_level2].unique():\n",
    "    plot_data = data_dinoflagellates[data_dinoflagellates[aggregation_level2] == station]\n",
    "    # Normalise mean of TPL values to the total mean TPL of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPL / plot_data.groupby('month').TPL.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "    \n",
    "    fig = px.histogram(plot_data, x='rel_expression_per_month', y=aggregation_level, color=tax_level,\n",
    "            color_discrete_map={\n",
    "                \"Rare\": \"#545454\",\n",
    "                \"Alexandrium\": \"#ebac23\",\n",
    "                \"Amphidinium\": \"#b80058\",\n",
    "                \"Noctiluca\": \"#008cf9\",\n",
    "                \"Tripos\": \"#006e00\",\n",
    "                \"Kryptoperidinium\": \"#00bbad\",\n",
    "                \"Oxyrrhis\": \"#d163e6\",\n",
    "                \"Karenia\": \"#b24502\",\n",
    "                \"Symbiodinium\": \"#ff9287\",\n",
    "                \"Ceratium\": \"#5954d6\",\n",
    "                \"Durinskia\": \"#00c6f8\",\n",
    "                \"Heterocapsa\": \"#878500\",\n",
    "                \"Scripsiella\": \"#00a76c\",\n",
    "                'Polarella': \"#FC8D62\",\n",
    "                'Gonyaulax': \"#8DA0CB\",\n",
    "                'Azadinium': \"#E78AC3\",\n",
    "                'Lingulodinium': \"#A6D854\",\n",
    "                'Karlodinium': \"#B3B3CC\"\n",
    "            },\n",
    "            category_orders={aggregation_level: month_order,\n",
    "                             tax_level: ['Rare', 'Noctiluca', 'Alexandrium', 'Durinskia', 'Kryptoperidinium', 'Gymnodinium', 'Pelagodinium', 'Polykrikos', 'Tripos']},\n",
    "            title=station)\n",
    "    \n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",\n",
    "            size=8,\n",
    "            color=\"#000000\"\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width=8.5 * pixels_per_cm,\n",
    "        height=7.5 * pixels_per_cm,\n",
    "        margin=dict(\n",
    "            l=0,\n",
    "            r=25,\n",
    "            b=25,\n",
    "            t=25\n",
    "        ),\n",
    "        xaxis_title_text='Month',\n",
    "        yaxis_title_text='Relative dinoflagellate TPL',\n",
    "        legend_title_text=tax_level\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as PNG and SVG\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_TPL_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_TPL_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_TPL_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_relative_TPL_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the taxonomic level and aggregation levels\n",
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "# Filter the data for dinoflagellates\n",
    "#data_dinoflagellates = data[(data['class'] == 'Dinophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPL']]\n",
    "\n",
    "# Remove transcripts below a certain TPL threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPL'] > 1]\n",
    "\n",
    "# Group by month, station, and species, sum TPL\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "# Specify the desired order of months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "               \"February_2021\", \"April_2021\", \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Plot total TPL per station\n",
    "pixels_per_cm = 37.79527559055118\n",
    "for station in data_dinoflagellates[aggregation_level2].unique():\n",
    "    plot_data = data_dinoflagellates[data_dinoflagellates[aggregation_level2] == station]\n",
    "    # Normalise mean of TPL values to the total mean TPL of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPL / plot_data.groupby('month').TPL.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "    \n",
    "    fig = px.histogram(plot_data, x='TPL', y=aggregation_level, color=tax_level,\n",
    "            color_discrete_map={\n",
    "                \"Rare\": \"#545454\",\n",
    "                \"Alexandrium\": \"#ebac23\",\n",
    "                \"Amphidinium\": \"#b80058\",\n",
    "                \"Noctiluca\": \"#008cf9\",\n",
    "                \"Tripos\": \"#006e00\",\n",
    "                \"Kryptoperidinium\": \"#00bbad\",\n",
    "                \"Oxyrrhis\": \"#d163e6\",\n",
    "                \"Karenia\": \"#b24502\",\n",
    "                \"Symbiodinium\": \"#ff9287\",\n",
    "                \"Ceratium\": \"#5954d6\",\n",
    "                \"Durinskia\": \"#00c6f8\",\n",
    "                \"Heterocapsa\": \"#878500\",\n",
    "                \"Scripsiella\": \"#00a76c\",\n",
    "                'Polarella': \"#FC8D62\",\n",
    "                'Gonyaulax': \"#8DA0CB\",\n",
    "                'Azadinium': \"#E78AC3\",\n",
    "                'Lingulodinium': \"#A6D854\",\n",
    "                'Karlodinium': \"#B3B3CC\"\n",
    "            },\n",
    "            category_orders={aggregation_level: month_order,\n",
    "                            tax_level: ['Rare', 'Noctiluca', 'Alexandrium', 'Durinskia', 'Kryptoperidinium', 'Gymnodinium', 'Pelagodinium', 'Polykrikos', 'Tripos']},\n",
    "            title=station)\n",
    "    \n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",\n",
    "            size=8,\n",
    "            color=\"#000000\"\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width=8.5 * pixels_per_cm,\n",
    "        height=7.5 * pixels_per_cm,\n",
    "        margin=dict(\n",
    "            l=0,\n",
    "            r=25,\n",
    "            b=25,\n",
    "            t=25\n",
    "        ),\n",
    "        xaxis_title_text='Month',\n",
    "        yaxis_title_text='Total dinoflagellate TPL',\n",
    "        legend_title_text=tax_level,\n",
    "        # Set range of x-axis\n",
    "        xaxis_range=[0, 700000000]\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as PNG and SVG\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPL_{}_at_{}_phylodb.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPL_{}_at_{}_phylodb.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPL_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_total_TPL_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Biodiversity indices\n",
    "Now, we'll generate samples x species tables and use those to generate biodiversity indices per sample. Then, we'll visualize patterns in said indices over time (in months)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the data\n",
    "## Get the diatoms\n",
    "# tax_level = 'genus'\n",
    "tax_level = 'Name_to_Use'\n",
    "aggregation_level = 'sample'\n",
    "\n",
    "## Since we'll be looking at the relative abundance of different diatom genera, we can only include reads that are annotated to a genus with a sufficient % sequence identity\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# First, I only want to include transcriptome bins that have at least 100 non-zero expressed transcripts in one of the samples\n",
    "# Group by month and genus, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, tax_level]).count().reset_index()\n",
    "\n",
    "# Now transform the data to the wide format\n",
    "data_dinoflagellates = data_dinoflagellates.pivot(index=tax_level, columns=aggregation_level, values='TPM')\n",
    "# Set NaN values to 0\n",
    "data_dinoflagellates = data_dinoflagellates.fillna(0)\n",
    "\n",
    "# Visualize the data\n",
    "data_dinoflagellates.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dinoflagellates[data_dinoflagellates > 100].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many diatom genera exist per month that have > 100 non-zero TPM transcripts\n",
    "## Per sample, count the number of genera that have > 100 TPM\n",
    "taxonomic_bin_abundance = data_dinoflagellates[data_dinoflagellates > 100].count(axis=0).reset_index()\n",
    "taxonomic_bin_abundance.columns = ['sample', 'num_species']\n",
    "\n",
    "## Add metadata\n",
    "taxonomic_bin_abundance = taxonomic_bin_abundance.merge(meta[['month']], left_on='sample', right_index=True, how='left')\n",
    "\n",
    "# Plot the number of diatom genera per month that have > 100 TPM\n",
    "\n",
    "fig = px.box(taxonomic_bin_abundance, x='month', y='num_species',\n",
    "             category_orders={'month': ['July_2020', \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "               \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "               \"May_2021\", \"June_2021\", \"July_2021\"]})\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save the figure\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/num_dinoflagellate_species_per_month.png\", width = 3.5, height = 2, scale=1, format='svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the genera that have > 100 TPM in at least one sample\n",
    "taxonomic_bin_abundance = data_dinoflagellates[data_dinoflagellates > 100].count(axis=1)\n",
    "taxonomic_bin_abundance.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot distribution of most abundant dinoflagellate genera\n",
    "taxonomic_bin_abundance.sort_values(ascending=False).plot(kind='bar', figsize=(10, 5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the genera that have > 100 TPM in at least one sample\n",
    "taxonomic_bin_abundance = data_dinoflagellates[data_dinoflagellates > 100].count(axis=1).reset_index()\n",
    "taxonomic_bin_abundance.columns = ['taxonomic_bin', 'abundant_in_samples']\n",
    "taxonomic_bin_abundance.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Since we'll be looking at the relative abundance of different diatom genera, we can only include reads that are annotated to a genus with a sufficient % sequence identity\n",
    "data_dinoflagellates = data[(data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])) & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# Group by month and genus, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([aggregation_level, tax_level]).count().reset_index()\n",
    "\n",
    "# Now transform the data to the wide format\n",
    "data_dinoflagellates = data_dinoflagellates.pivot(index=tax_level, columns=aggregation_level, values='TPM')\n",
    "# Set NaN values to 0\n",
    "data_dinoflagellates = data_dinoflagellates.fillna(0)\n",
    "\n",
    "# Filter the dataframe for these genera\n",
    "abundant_taxa = taxonomic_bin_abundance[taxonomic_bin_abundance['abundant_in_samples'] >= 1]['taxonomic_bin'].tolist()\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates.index.isin(abundant_taxa)]\n",
    "\n",
    "# Save to csv\n",
    "data_dinoflagellates.to_csv(f'../../data/analysis/dinoflagellate_{tax_level}_{aggregation_level}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now save the count of these genera per sample\n",
    "taxonomic_bin_abundance = data_dinoflagellates[data_dinoflagellates > 100].count(axis=0).reset_index()\n",
    "taxonomic_bin_abundance.columns = ['sample', 'num_species']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomic_bin_abundance.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now run the above through the R [diversity](diversity.r) script. Then, we'll visualize the results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load biodiversity estimates\n",
    "dinoflagellate_biodiversity = pd.read_csv('../../data/analysis/dinoflagellates_biodiversity_estimates.csv', index_col=0)\n",
    "\n",
    "# Add metadata to the biodiversity estimates\n",
    "meta = pd.read_csv('../../samples.csv', sep=';', index_col=0)\n",
    "dinoflagellate_biodiversity = dinoflagellate_biodiversity.merge(meta, left_index=True, right_index=True)\n",
    "\n",
    "# Add the number of dinoflagellate species per sample\n",
    "dinoflagellate_biodiversity = dinoflagellate_biodiversity.merge(taxonomic_bin_abundance, left_index=True, right_on='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot number of dinoflagellate genera per month\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(5*cm, 5*cm))\n",
    "\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxplot(x=\"month\", y=\"num_species\",\n",
    "              order=month_order,\n",
    "              data=dinoflagellate_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Change x axis labels rotation\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/dinoflagellate_genera_per_month.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot evenness per month\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(5*cm, 5*cm))\n",
    "\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxenplot(x=\"month\", y=\"evenness\",\n",
    "              order=month_order,\n",
    "              scale=\"linear\", data=dinoflagellate_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Change x axis labels rotation\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "\n",
    "# Set y axis limits\n",
    "plt.ylim(-0.05, 1.05)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/dinoflagellate_evenness_per_month.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Calculate Functional Richness"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Data preparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set aggregation levels of interest\n",
    "functional_info = 'KEGG_ko'\n",
    "tax_level = 'Name_to_Use'\n",
    "# tax_level = 'genus'\n",
    "aggregation_level = 'sample'\n",
    "# Only retain abundant species determined above\n",
    "data_dinoflagellates = data[data[tax_level].isin(abundant_taxa)][[aggregation_level, functional_info, 'TPM']]\n",
    "\n",
    "# if processing kegg data, extra preprocessing is required: \n",
    "# Comment or uncomment the following line if multiple values are assigned to a transcript!\n",
    "## split them up\n",
    "data_dinoflagellates = data_dinoflagellates.assign(**{functional_info:data_dinoflagellates[functional_info].str.split(',')})\n",
    "# Now we can explode the functional column\n",
    "data_dinoflagellates = data_dinoflagellates.explode(functional_info)\n",
    "# Cut off that weird part of the identifier, if necessary!\n",
    "data_dinoflagellates[functional_info] = data_dinoflagellates[functional_info].str.split(\":\", expand=True).drop(columns=0)\n",
    "\n",
    "# Remove the rows with no functional annotation\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates[functional_info].notna()]\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates[functional_info] != '-']\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates['TPM'] > 1]\n",
    "\n",
    "# Group by functional information and sample values, sum TPM\n",
    "functional_diversity = data_dinoflagellates.groupby([functional_info, aggregation_level]).sum().reset_index()\n",
    "\n",
    "# Reshape the dataframe to wide format\n",
    "functional_diversity = functional_diversity.pivot(index=functional_info, columns=aggregation_level, values='TPM')\n",
    "\n",
    "# Set NaN values to 0\n",
    "functional_diversity = functional_diversity.fillna(0)\n",
    "\n",
    "# Calculate functional richness\n",
    "functional_diversity[functional_diversity < 1] = 0\n",
    "functional_diversity[functional_diversity >= 1] = 1\n",
    "\n",
    "functional_richness =  functional_diversity.sum(axis=0).reset_index()\n",
    "\n",
    "functional_richness.columns = ['sample', functional_info]\n",
    "functional_richness.set_index('sample', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add the functional richness to the biodiversity estimates\n",
    "dinoflagellate_biodiversity = dinoflagellate_biodiversity.merge(functional_richness, left_on='sample', right_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dinoflagellate_biodiversity.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Abundance of functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the amount of KO per month in a boxplot\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(5*cm, 5*cm))\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxenplot(x=\"month\", y=\"KEGG_ko\",\n",
    "              order=month_order,\n",
    "              scale=\"linear\", data=dinoflagellate_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Change x axis labels rotation\n",
    "plt.xticks(rotation=90, ha='right')\n",
    "\n",
    "# Set y axis limits\n",
    "plt.ylim(-50, 5500)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/dinoflagellates_KO_abundance_per_month.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# How is functional richness correlated to the amount of diatom reads?\n",
    "## Calculate the number of reads that are assigned to diatoms in a given sample\n",
    "number_of_reads = data[data['Taxogroup2_UniEuk'] == 'Dinophyceae'][[aggregation_level, 'TPM']]\n",
    "number_of_reads = number_of_reads.groupby(aggregation_level).sum().reset_index()\n",
    "\n",
    "# Add it to the diatom_biodiversity data\n",
    "dinoflagellate_biodiversity = dinoflagellate_biodiversity.merge(number_of_reads, on='sample')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot\n",
    "fig = px.scatter(dinoflagellate_biodiversity, x=\"TPM\", y=\"KEGG_ko\")\n",
    "\n",
    "fig.update_traces(marker=dict(size=5,\n",
    "                              color='#63C5DA',\n",
    "                              line=dict(width=1,\n",
    "                                        color='DarkSlateGrey')),\n",
    "                  selector=dict(mode='markers'))\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 5.5 * pixels_per_cm,\n",
    "    height= 5.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    plot_bgcolor='white',\n",
    "    showlegend=False\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Active species richness vs abundance of functional modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_data = dinoflagellate_biodiversity.copy()\n",
    "\n",
    "# Create a linear model\n",
    "linear_model = LinearRegression()\n",
    "# Create a polynomial model\n",
    "poly_model = make_pipeline(PolynomialFeatures(degree=2), LinearRegression())\n",
    "\n",
    "# Calculate the cross-validated scores for all models using mean squared error as the scoring metric\n",
    "linear_scores = cross_val_score(linear_model, plot_data[['num_species']], plot_data['KEGG_ko'], cv=5)\n",
    "poly_scores = cross_val_score(poly_model, plot_data[['num_species']], plot_data['KEGG_ko'], cv=5)\n",
    "\n",
    "# Compare the average scores of both models\n",
    "print(\"Linear model: %0.2f accuracy with a standard deviation of %0.2f\" % (linear_scores.mean(), linear_scores.std()))\n",
    "print(\"Polynomial model: %0.2f accuracy with a standard deviation of %0.2f\" % (poly_scores.mean(), poly_scores.std()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Polynomial model is the better fit\n",
    "## Fit the polynomial model\n",
    "poly_model.fit(plot_data[['num_species']], plot_data['KEGG_ko'])\n",
    "## Extract the predictions of the polynomial model\n",
    "# First generate x-values for the polynomial fit\n",
    "x_values = np.linspace(plot_data['num_species'].min(), plot_data['num_species'].max(), 100)\n",
    "\n",
    "# Predict y-values using the polynomial fit\n",
    "model_predictions = poly_model.predict(x_values.reshape(-1, 1))\n",
    "\n",
    "# Set conversion factor to transform pixels to cm\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "fig = px.scatter(plot_data, x=\"num_species\", y=\"KEGG_ko\")\n",
    "\n",
    "# Add the polynomial fit\n",
    "fig.add_trace(go.Scatter(\n",
    "    x=x_values,\n",
    "    y=model_predictions,\n",
    "    mode='lines',\n",
    "    line=dict(color='black', width=1),\n",
    "    name='Polynomial Fit'\n",
    "))\n",
    "\n",
    "fig.update_traces(marker=dict(size=5,\n",
    "                              color='#63C5DA',\n",
    "                              line=dict(width=1,\n",
    "                                        color='DarkSlateGrey')),\n",
    "                  selector=dict(mode='markers'))\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 5.5 * pixels_per_cm,\n",
    "    height= 5.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    plot_bgcolor='white',\n",
    "    showlegend=False\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    linewidth=0.5,\n",
    "    gridcolor='lightgrey',\n",
    "    range = [-50, 5500]\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# There's fewer active species in dinoflagellates, so let's just create a boxplot of the number of active species vs functional richness\n",
    "# Plot the amount of KO per month in a boxplot\n",
    "# Set figure size\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(5*cm, 5*cm))\n",
    "\n",
    "# Create a plot of the functional annotation per month\n",
    "sns.boxenplot(x=\"num_species\", y=\"KEGG_ko\",\n",
    "              scale=\"linear\", data=dinoflagellate_biodiversity,\n",
    "              color='#63C5DA')\n",
    "\n",
    "# Set y axis limits\n",
    "plt.ylim(-50, 5500)\n",
    "\n",
    "# Add grid\n",
    "plt.grid(axis='both')\n",
    "\n",
    "# save figure as svg\n",
    "plt.savefig(f'../../figures/diatoms_vs_dinoflagellates/dinoflagellates_KO_abundance_num_species.svg', dpi=600, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Colour according to month\n",
    "fig = px.scatter(plot_data, x=\"num_species\", y=\"KEGG_ko\", trendline=\"ols\", color='month', color_discrete_sequence=px.colors.qualitative.Set1)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=14,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 14 * pixels_per_cm,\n",
    "    height= 10 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    plot_bgcolor='white',\n",
    ")\n",
    "\n",
    "fig.update_xaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    gridcolor='lightgrey'\n",
    ")\n",
    "fig.update_yaxes(\n",
    "    mirror=True,\n",
    "    ticks='outside',\n",
    "    showline=True,\n",
    "    linecolor='black',\n",
    "    gridcolor='lightgrey',\n",
    "    range = [-50, 5500]\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save figure as png\n",
    "fig.write_image(f'../../figures/diatoms_vs_dinoflagellates/dinoflagellate_KO_abundance_vs_species_per_month.svg')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trophic feeding mode PFAMs\n",
    "From the above we can hypothesize that the dinoflagellate community is mixotrophic. We'll see if this is reflected in the expression of feeding-mode related PFAMS."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, create a new data_diato dataframe with PFAM expression data\n",
    "functional_info = 'PFAMs'\n",
    "aggregation_level = 'month'\n",
    "data_dinoflagellates_PFAM = data[data['class'] == 'Dinophyceae'][[aggregation_level, functional_info, 'TPM']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many of the relevant PFAMs are present in the data\n",
    "data_dinoflagellates_PFAM[functional_info].isin(relevant_pfams['short_name']).value_counts(normalize=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check to which clade these hits belong\n",
    "relevant_pfams[relevant_pfams['short_name'].isin(data_dinoflagellates_PFAM[functional_info])]['clade'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only retain the relevant PFAMs\n",
    "data_dinoflagellates_PFAM = data_dinoflagellates_PFAM[data_dinoflagellates_PFAM[functional_info].isin(relevant_pfams['short_name'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What are the most abundant PFAMs?\n",
    "data_dinoflagellates_PFAM.groupby(functional_info)['TPM'].sum().sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dinoflagellates_PFAM.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Determine the genes with highest variance across samples\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = data_dinoflagellates_PFAM.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Add number of stations visited per month\n",
    "samples_per_month = {\n",
    "    'July_2020': 6,\n",
    "    'August_2020': 6,\n",
    "    'September_2020': 6,\n",
    "    'November_2020': 6,\n",
    "    'December_2020': 6,\n",
    "    'January_2021': 5,\n",
    "    'February_2021': 5,\n",
    "    'April_2021': 4,\n",
    "    'May_2021': 6,\n",
    "    'June_2021': 6,\n",
    "    'July_2021': 6\n",
    "}\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to dinoflagellates\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "# Re-order the columns according to month\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# In the index, include the PFAM clade and functional descriptions to the short names between brackets\n",
    "plot_data_scaled.index = plot_data_scaled.index.map(lambda x: \"{} ({} - {})\".format(x, relevant_pfams[relevant_pfams['short_name'] == x]['clade'].values[0], relevant_pfams[relevant_pfams['short_name'] == x]['function'].values[0]))\n",
    "\n",
    "# Only retain the 15 most variable PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.var(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_data_scaled.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the heatmap\n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False,\n",
    "    # Legend bar should be on the right\n",
    "    cbar_kws={'orientation': 'vertical'},\n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Save figure as svg\n",
    "g.savefig(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_heatmap_{}_most_variable_trophic_{}_per_{}.svg\".format(number_of_genes, functional_info, aggregation_level), format='svg', dpi=600)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'll do the same but with the 15 most expressed PFAMs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Determine the genes with highest variance across samples\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = data_dinoflagellates_PFAM.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Add number of stations visited per month\n",
    "samples_per_month = {\n",
    "    'July_2020': 6,\n",
    "    'August_2020': 6,\n",
    "    'September_2020': 6,\n",
    "    'November_2020': 6,\n",
    "    'December_2020': 6,\n",
    "    'January_2021': 5,\n",
    "    'February_2021': 5,\n",
    "    'April_2021': 4,\n",
    "    'May_2021': 6,\n",
    "    'June_2021': 6,\n",
    "    'July_2021': 6\n",
    "}\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to dinoflagellates\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "# Re-order the columns according to month\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# In the index, include the PFAM clade and functional descriptions to the short names between brackets\n",
    "plot_data_scaled.index = plot_data_scaled.index.map(lambda x: \"{} ({} - {})\".format(x, relevant_pfams[relevant_pfams['short_name'] == x]['clade'].values[0], relevant_pfams[relevant_pfams['short_name'] == x]['function'].values[0]))\n",
    "\n",
    "# Only retain the 15 most expressed PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.sum(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]\n",
    "\n",
    "# \n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False,\n",
    "    # Legend bar should be on the right\n",
    "    cbar_kws={'orientation': 'vertical'},\n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Save figure as svg\n",
    "g.savefig(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_heatmap_{}_most_expressed_trophic_{}_per_{}.svg\".format(number_of_genes, functional_info, aggregation_level), format='svg', dpi=600)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'll do the same with the full feature set (all 1046 gene families used in the model described in the paper)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PFAM_features = pd.read_excel('../../data/external/pnas.2100916119.sd03.xlsx', skiprows=1, header=1)\n",
    "\n",
    "PFAM_features = PFAM_features.dropna()\n",
    "\n",
    "PFAM_features.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Of the PFAM features, which are most expressed?\n",
    "## Subset the data to only include the PFAMs\n",
    "functional_info = 'PFAMs'\n",
    "aggregation_level = 'month'\n",
    "data_dinoflagellates_PFAM = data[data['class'] == 'Dinophyceae'][[aggregation_level, functional_info, 'TPM']]\n",
    "dinoflagellate_features = data_dinoflagellates_PFAM[data_dinoflagellates_PFAM[functional_info].isin(PFAM_features['Name'])]\n",
    "\n",
    "# Group by month, PFAM and sum the TPMs\n",
    "dinoflagellate_features = dinoflagellate_features.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Print the top 10 most expressed PFAMs \n",
    "dinoflagellate_features.groupby(functional_info)['TPM'].sum().sort_values(ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualize the most abundant PFAMs\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = dinoflagellate_features.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to dinoflagellates\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# Only retain the 15 most variable PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.median(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the heatmap\n",
    "sns.set()\n",
    "\n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False\n",
    "    \n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now we'll do the same but with the 15 most highly expressed PFAMs\n",
    "# Visualize the most abundant PFAMs\n",
    "number_of_PFAMs = 15\n",
    "# Log2 transform the data\n",
    "plot_data = dinoflagellate_features.groupby([aggregation_level, functional_info])['TPM'].sum().reset_index()\n",
    "\n",
    "# Divide the TPM sum by number of samples per month to obtain the average TPM per month\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data[aggregation_level].map(samples_per_month)\n",
    "\n",
    "# Divide the TPM sum by the total TPM sum per month dedicated to dinoflagellates\n",
    "plot_data['TPM'] = plot_data['TPM'] / plot_data.groupby(aggregation_level)['TPM'].transform('sum')\n",
    "\n",
    "# Swap the short PFAM names for their functional descriptions\n",
    "\n",
    "# Transform the data to the wide format\n",
    "plot_data_wide = plot_data.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "# log2 transform the data\n",
    "plot_data_log = np.log2(plot_data_wide + 1)\n",
    "\n",
    "# Using non-log2 transformed data gives more weight to more expressed genes; these have higher variances\n",
    "# Using log2 transformed data gives more weight to less expressed genes; and counterselects for highly expressed genes\n",
    "\n",
    "# Transpose the data\n",
    "print(plot_data_log.shape)\n",
    "plot_data_log = plot_data_log.transpose()\n",
    "\n",
    "plot_data_log = plot_data_log.reindex(month_order, axis=1)\n",
    "\n",
    "# Scale the data\n",
    "plot_data_scaled = plot_data_log.apply(lambda x: (x - x.mean()) / x.std(), axis=1)\n",
    "\n",
    "# Cluster similar rows together\n",
    "plot_data_scaled = plot_data_scaled.reindex(plot_data_scaled.mean(axis=1).sort_values(ascending=False).index, axis=0)\n",
    "\n",
    "# Only retain the 15 most expressed PFAMs\n",
    "most_variable_PFAMs = plot_data_scaled.sum(axis=1).sort_values(ascending=False).head(number_of_PFAMs).index\n",
    "plot_data_scaled = plot_data_scaled.loc[most_variable_PFAMs]\n",
    "\n",
    "# \n",
    "g = sns.clustermap(\n",
    "    plot_data_scaled,\n",
    "    figsize=(8, 6),\n",
    "    cmap='mako',\n",
    "    col_cluster=False,\n",
    "    # Legend bar should be on the right\n",
    "    cbar_kws={'orientation': 'vertical'},\n",
    "    )\n",
    "\n",
    "g.ax_heatmap.set_xticklabels(g.ax_heatmap.get_xticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_yticklabels(g.ax_heatmap.get_yticklabels(), fontsize=8)\n",
    "g.ax_heatmap.set_facecolor('#f7f7f7')\n",
    "g.ax_heatmap.set_xlabel(None)\n",
    "g.ax_heatmap.set_ylabel(None)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Save figure as svg\n",
    "# g.savefig(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellates_heatmap_{}_most_variable_trophic_{}_per_{}.svg\".format(number_of_genes, functional_info, aggregation_level), format='svg', dpi=600)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trophic feeding mode of dinoflagellates\n",
    "I've predicted the trophic feeding mode of several species using the machine learning model of the paper by [Lambert & Groussman, 2022](https://doi.org/10.1073/pnas.2100916119), see [here](./trophic_mode_prediction.ipynb). We'll use the data to see how the trophic feeding mode of dinoflagellates changes over time, same as for the diatoms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "#trophic_predictions = pd.read_csv('../../data/analysis/phylodb_trophic_mode_predictions.csv')\n",
    "trophic_predictions = pd.read_csv('../../data/analysis/eukprot_trophic_mode_predictions.csv')\n",
    "# Only retain the diatoms\n",
    "#trophic_predictions = trophic_predictions[trophic_predictions['class'] == 'Dinophyceae']\n",
    "trophic_predictions = trophic_predictions[trophic_predictions['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])]\n",
    "print(len(trophic_predictions))\n",
    "trophic_predictions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the relative abundance of the three trophic modes per month\n",
    "# Calculate relative abundance of each prediction per month\n",
    "rel_abundance = (trophic_predictions[['month', 'station', 'prediction']].groupby(['month', 'prediction'])\n",
    "                       .size()\n",
    "                       .groupby(level=0).apply(lambda x: 100 * x / x.sum())\n",
    "                       .reset_index(name='relative abundance'))\n",
    "# Set the color palette\n",
    "color_map = {\n",
    "    'Phot': 'green',\n",
    "    'Mix': 'black',\n",
    "    'Het': 'red'\n",
    "}\n",
    "fig = px.histogram(rel_abundance.sort_values(\"month\", ascending=False),\n",
    "             x = \"relative abundance\",\n",
    "             y = \"month\",\n",
    "             color = \"prediction\",\n",
    "             orientation='h',\n",
    "             color_discrete_map=color_map,\n",
    "             category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"]})\n",
    "\n",
    "pixels_per_cm = 37.7952755906\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8  # Set the font size\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Ratio of predicted trophic modes (%)',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellate_trophic_mode_relative_abundance.svg\")\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellate_trophic_mode_relative_abundance_eukprot.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the relevant columns\n",
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "df = trophic_predictions[[tax_level, 'month', 'prediction']]\n",
    "# Convert 'prediction' into numerical categories\n",
    "df['prediction'] = df['prediction'].map({'Phot':1, 'Mix':2, 'Het':3})\n",
    "\n",
    "# Pivot the DataFrame to use for heatmap\n",
    "df_pivot = df.pivot_table(index=tax_level, columns='month', values='prediction', \n",
    "                          aggfunc=lambda x: int(stats.mode(x[x!=0])[0][0] if any(x!=0) else 0))\n",
    "\n",
    "# Fill NA with a specific category (e.g., 0 for 'no prediction')\n",
    "df_pivot = df_pivot.fillna(0)\n",
    "\n",
    "# Ensure the values are integer\n",
    "df_pivot = df_pivot.astype(int)\n",
    "\n",
    "# Set the order of the months\n",
    "df_pivot = df_pivot.reindex(month_order, axis=1)\n",
    "\n",
    "# Set the order of the genera to be the same as in the above relative abundance plot\n",
    "df_pivot = df_pivot.reindex(dinoflagellate_legend_order, axis=0)\n",
    "\n",
    "# Create a color map\n",
    "cmap = mcolors.ListedColormap(['lightgrey', 'green', 'yellow', 'red'])\n",
    "\n",
    "# Create the heatmap with the color map\n",
    "plt.figure(figsize=(4.5, 4))\n",
    "sns.heatmap(df_pivot, cmap=cmap, annot=False, cbar=False, vmin=0, vmax=3)\n",
    "\n",
    "# Change layout to match the relative abundance plot\n",
    "plt.tight_layout()\n",
    "\n",
    "# Remove the x and  y-axis label\n",
    "plt.xlabel(None)\n",
    "plt.ylabel(None)\n",
    "\n",
    "# Rotate the -axis labels\n",
    "plt.yticks(rotation=0)\n",
    "\n",
    "# Font size should be 8, and font family Times New Roman\n",
    "plt.rcParams.update({'font.size': 8, 'font.family': 'Times New Roman'})\n",
    "\n",
    "# Save figure as svg\n",
    "#plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellate_trophic_mode_genus_consensus_heatmap.svg\", format='svg', dpi=600)\n",
    "plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellate_trophic_mode_genus_consensus_heatmap_eukprot.svg\", format='svg', dpi=600)\n",
    "\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df[df[tax_level] == 'Noctiluca']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify how the matrix will look like (in the end its aggregation_level x functional_info)\n",
    "functional_info = 'KEGG_ko'\n",
    "aggregation_level = 'sample'\n",
    "\n",
    "# Prepare gene expression data and load environmental variables\n",
    "#data_dinoflagellates = data[data['class'] == 'Dinophyceae'][[aggregation_level, functional_info, 'TPM']]\n",
    "data_dinoflagellates = data[data['Taxogroup2_UniEuk'].isin(['Dinophyceae', 'core-Noctilucales'])][[aggregation_level, functional_info, 'TPM']]\n",
    "\n",
    "# if processing kegg data, extra preprocessing is required: \n",
    "# Comment or uncomment the following line if multiple values are assigned to a transcript!\n",
    "## split KEGG identifiers up\n",
    "data_dinoflagellates = data_dinoflagellates.assign(**{functional_info:data_dinoflagellates[functional_info].str.split(',')})\n",
    "# Now we can explode the functional column\n",
    "data_dinoflagellates = data_dinoflagellates.explode(functional_info)\n",
    "# Cut off that weird part of the identifier, if necessary!\n",
    "data_dinoflagellates[functional_info] = data_dinoflagellates[functional_info].str.split(\":\", expand=True).drop(columns=0)\n",
    "\n",
    "# Group by functional information and sample values, sum TPM\n",
    "data_dinoflagellates = data_dinoflagellates.groupby([functional_info, aggregation_level]).sum().reset_index()\n",
    "\n",
    "# Remove the rows with no functional annotation\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates[functional_info].notna()]\n",
    "data_dinoflagellates = data_dinoflagellates[data_dinoflagellates[functional_info] != '-']\n",
    "\n",
    "# Transform the data to the wide format\n",
    "data_dinoflagellates = data_dinoflagellates.pivot(index=aggregation_level, columns=functional_info, values='TPM')\n",
    "\n",
    "# Log2 transform the data\n",
    "data_dinoflagellates_log = np.log2(data_dinoflagellates + 1)\n",
    "\n",
    "# Scale the features (columns, TPM values of every prediction)\n",
    "data_dinoflagellates_log = pd.DataFrame(StandardScaler().fit_transform(data_dinoflagellates_log),\n",
    "                                   index=data_dinoflagellates.index,\n",
    "                                   columns=data_dinoflagellates.columns)\n",
    "## Scaling removes the mean and scales to unit variance, the resulting values are z-scores\n",
    "\n",
    "# Load the environmental variables\n",
    "env_variables = pd.read_csv(\"../../data/environmental/samples_env.csv\", sep=\";\")\n",
    "env_variables = env_variables.set_index('sample')\n",
    "\n",
    "# Merge environmental variables with gene expression data\n",
    "pca_data = data_dinoflagellates_scaled.merge(env_variables, left_index=True, right_on='sample', how='left')\n",
    "\n",
    "# Perform PCA on gene expression data, ignore columns that are also in the environmental variables\n",
    "pca = PCA(n_components=2)\n",
    "principalComponents = pca.fit_transform(pca_data.drop(columns=env_variables.columns))\n",
    "\n",
    "# Add principal components to the data\n",
    "pca_data['PC1'] = principalComponents[:, 0]\n",
    "pca_data['PC2'] = principalComponents[:, 1]\n",
    "\n",
    "# Create a color dictionary for the months\n",
    "month_color_dict = dict(zip(data['month'].unique(), sns.color_palette('tab10', n_colors=len(data['month'].unique()))))\n",
    "\n",
    "# Set figure size and font scale\n",
    "cm = 1/2.54\n",
    "plt.figure(figsize=(12*cm, 14*cm))\n",
    "sns.set(style='white', font_scale=1)\n",
    "\n",
    "# Plot PCA biplot with the color of the month corrresponding to the month of the sample\n",
    "# and the shape of the point corresponding to the station of the sample\n",
    "sns.scatterplot(data=pca_data,\n",
    "                x='PC1',\n",
    "                y='PC2',\n",
    "                hue='month',\n",
    "                hue_order=month_order,\n",
    "                style='station',\n",
    "                palette=month_color_dict,\n",
    "                s=40, edgecolor='black', linewidth=0.5)\n",
    "\n",
    "plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left', borderaxespad=0)\n",
    "# Plot labels with the variance explained by each principal component\n",
    "plt.xlabel('Principal Component 1 ({}%)'.format(round(pca.explained_variance_ratio_[0] * 100, 2)))\n",
    "plt.ylabel('Principal Component 2 ({}%)'.format(round(pca.explained_variance_ratio_[1] * 100, 2)))\n",
    "\n",
    "# Plot arrows indicating the correlation between the principal components and the environmental variables\n",
    "## Linear regression to determine the correlation between the principal components and the environmental variables\n",
    "columns_of_interest = ['NO3', 'PO4', 'Si', 'SPM', 'salinity', 'Temperature']\n",
    "\n",
    "# Calculate the correlation matrix between the principal components and the environmental variables\n",
    "corr_matrix = np.corrcoef(pca_data[['PC1', 'PC2'] + columns_of_interest].dropna().T)\n",
    "\n",
    "# Get the correlation between the first two principal components and each environmental parameter\n",
    "corr_PC1_env = corr_matrix[0, 2:]\n",
    "corr_PC2_env = corr_matrix[1, 2:]\n",
    "\n",
    "# Plot the arrows representing the correlations\n",
    "for i, env_param in enumerate(columns_of_interest):\n",
    "    plt.arrow(0, 0, corr_PC1_env[i]*max(pca_data['PC1']), corr_PC2_env[i]*max(pca_data['PC2']), head_width=0.05, color='gray')\n",
    "    # corr_PC1_env[i] specifies the end coordinate of the arrow in the x direction\n",
    "    # This coordinate is multiplied with max(pca_data['PC1']) to scale the arrow to the length of the principal component for visualization\n",
    "    plt.text(corr_PC1_env[i]*max(pca_data['PC1'])*1.05, corr_PC2_env[i]*max(pca_data['PC2'])*1.05, env_param, fontsize=12, color='gray')\n",
    "    # Same as above, but the text is placed 5% further away from the arrow\n",
    "\n",
    "# Save figure as svg\n",
    "#plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellate_PCA_{}_per_{}.svg\".format(functional_info, aggregation_level), format='svg', dpi=600)\n",
    "plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/dinoflagellate_PCA_{}_per_{}_eukprot.svg\".format(functional_info, aggregation_level), format='svg', dpi=600)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Prymnesiophytes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, let's get a list of all the unique dinoflagellate species\n",
    "#prymnesiophytes = data[(data['class'] == 'Prymnesiophyceae')  & (data['p_ident'] >= 0.98)]\n",
    "data[(data['Taxogroup2_UniEuk'] == 'Prymnesiophyceae')  & (data['p_ident'] >= 0.98)]['Name_to_Use'].unique()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 4.1 Prymnesiophyte abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'species'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "\n",
    "#data_prymnesiophytes = data[(data['class'] == 'Prymnesiophyceae') & (data['genus'] == 'Phaeocystis') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_prymnesiophytes = data[(data['Taxogroup2_UniEuk'] == 'Prymnesiophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_prymnesiophytes = data_prymnesiophytes[data_prymnesiophytes['TPM'] > 1]\n",
    "\n",
    "# Group by month and species, sum TPM\n",
    "data_prymnesiophytes = data_prymnesiophytes.groupby([aggregation_level, tax_level]).sum().reset_index()\n",
    "\n",
    "data_prymnesiophytes['month'] = pd.Categorical(data_prymnesiophytes['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Normalise mean of TPM values to the total mean TPM of that month\n",
    "data_prymnesiophytes[\"rel_expression_per_month\"] = data_prymnesiophytes.TPM / data_prymnesiophytes.groupby('month').TPM.transform('sum')\n",
    "\n",
    "# Combine low-abundant groups\n",
    "rare_groups = data_prymnesiophytes['rel_expression_per_month'] <= 0.02\n",
    "data_prymnesiophytes.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "# Print unique species\n",
    "print(data_prymnesiophytes[tax_level].unique())\n",
    "# Inspect data\n",
    "data_prymnesiophytes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add conversion variable\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "# Plot\n",
    "fig = px.histogram(data_prymnesiophytes.sort_values(\"month\", ascending=False), \n",
    "                x=\"rel_expression_per_month\", \n",
    "                y=\"month\", \n",
    "                color=tax_level,\n",
    "                orientation='h',\n",
    "                color_discrete_map={\n",
    "                    \"Rare\": \"#545454\",\n",
    "                    \"Phaeocystis\": \"#F26B38\",\n",
    "                 },\n",
    "                # Always use the same order for the species\n",
    "                category_orders={tax_level: [\"Rare\", \"Phaeocystis\"],\n",
    "                                'month': [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                                            \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                                            \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                                            \"June_2021\", \"July_2021\"]}\n",
    "                # text_auto='.2f'\n",
    "                )\n",
    "\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\",  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='% TPM of total sum',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save figure as png\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_relative_expression_per_month_{}.png\".format(tax_level), scale=1)\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_relative_expression_per_month_{}.svg\".format(tax_level), scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_relative_expression_per_month_{}_eukprot.png\".format(tax_level), scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_relative_expression_per_month_{}_eukprot.svg\".format(tax_level), scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store the order of the genera in the legend\n",
    "prymnesiophyte_legend_order = []\n",
    "for i in range(len(fig.data)):\n",
    "    prymnesiophyte_legend_order.append(fig.data[i].name)\n",
    "print(prymnesiophyte_legend_order)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cool! Now let's only look at the Phaeocystis species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'species'\n",
    "tax_level = 'Name_to_Use'\n",
    "aggregation_level = 'month'\n",
    "\n",
    "#data_prymnesiophytes = data[(data['class'] == 'Prymnesiophyceae') & (data['genus'] == 'Phaeocystis') & (data['p_ident'] >= 0.98)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_prymnesiophytes = data[(data['Taxogroup2_UniEuk'] == 'Prymnesiophyceae') & (data['Genus_UniEuk'] == 'Phaeocystis') & (data['p_ident'] >= 0.98)][[aggregation_level, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_prymnesiophytes = data_prymnesiophytes[data_prymnesiophytes['TPM'] > 1]\n",
    "\n",
    "# Group by month and species, sum TPM\n",
    "data_prymnesiophytes = data_prymnesiophytes.groupby([aggregation_level, tax_level]).sum().reset_index()\n",
    "\n",
    "data_prymnesiophytes['month'] = pd.Categorical(data_prymnesiophytes['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Normalise mean of TPM values to the total mean TPM of that month\n",
    "data_prymnesiophytes[\"rel_expression_per_month\"] = data_prymnesiophytes.TPM / data_prymnesiophytes.groupby('month').TPM.transform('sum')\n",
    "\n",
    "# Combine low-abundant groups\n",
    "rare_groups = data_prymnesiophytes['rel_expression_per_month'] <= 0.02\n",
    "data_prymnesiophytes.loc[rare_groups, tax_level] = 'Rare'\n",
    "\n",
    "# Print unique species\n",
    "print(data_prymnesiophytes[tax_level].unique())\n",
    "# Inspect data\n",
    "data_prymnesiophytes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add conversion variable\n",
    "pixels_per_cm = 37.79527559055118\n",
    "\n",
    "# Plot\n",
    "fig = px.histogram(data_prymnesiophytes.sort_values(\"month\", ascending=False), \n",
    "                x=\"rel_expression_per_month\", \n",
    "                y=\"month\", \n",
    "                color=tax_level,\n",
    "                orientation='h',\n",
    "                color_discrete_map={\n",
    "                    \"Rare\": \"#545454\",\n",
    "                    \"Phaeocystis\": \"#F26B38\",\n",
    "                 },\n",
    "                # Always use the same order for the species\n",
    "                category_orders={tax_level: [\"Rare\", \"Phaeocystis\"],\n",
    "                                'month': [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                                            \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                                            \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                                            \"June_2021\", \"July_2021\"]}\n",
    "                # text_auto='.2f'\n",
    "                )\n",
    "\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\",  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='% TPM of total sum',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save figure as png\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_phaeocystis_relative_expression_per_month_{}.png\".format(tax_level), scale=1)\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_phaeocystis_relative_expression_per_month_{}.svg\".format(tax_level), scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_phaeocystis_relative_expression_per_month_{}_eukprot.png\".format(tax_level), scale=1)\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_phaeocystis_relative_expression_per_month_{}_eukprot.svg\".format(tax_level), scale=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Spatial distribution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "aggregation_level2 = 'station'\n",
    "\n",
    "#data_prymnesiophytes = data[(data['class'] == 'Prymnesiophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "data_prymnesiophytes = data[(data['Taxogroup2_UniEuk'] == 'Prymnesiophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, aggregation_level2, tax_level, 'TPM']]\n",
    "\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_prymnesiophytes = data_prymnesiophytes[data_prymnesiophytes['TPM'] > 1]\n",
    "\n",
    "# Group by month, station and species, sum TPM\n",
    "data_prymnesiophytes = data_prymnesiophytes.groupby([aggregation_level, aggregation_level2, tax_level]).sum().reset_index()\n",
    "\n",
    "data_prymnesiophytes['month'] = pd.Categorical(data_prymnesiophytes['month'], [\"July_2020\", \"August_2020\", \"September_2020\", \n",
    "                                                        \"November_2020\", \"December_2020\", \"January_2021\", \n",
    "                                                        \"February_2021\", \"April_2021\", \"May_2021\", \n",
    "                                                        \"June_2021\", \"July_2021\"])\n",
    "\n",
    "# Plot per station\n",
    "for station in data_prymnesiophytes[aggregation_level2].unique():\n",
    "    plot_data = data_prymnesiophytes[data_prymnesiophytes[aggregation_level2] == station]\n",
    "    # Normalise mean of TPM values to the total mean TPM of that month\n",
    "    plot_data[\"rel_expression_per_month\"] = plot_data.TPM / plot_data.groupby('month').TPM.transform('sum')\n",
    "\n",
    "    # Combine low-abundant groups\n",
    "    rare_groups = plot_data['rel_expression_per_month'] <= 0.02\n",
    "    plot_data.loc[rare_groups, tax_level] = 'Rare'\n",
    "    fig = px.histogram(plot_data.sort_values(\"month\", ascending=False), \n",
    "                    x=\"rel_expression_per_month\", \n",
    "                    y=\"month\", \n",
    "                    color=tax_level,\n",
    "                    orientation='h',\n",
    "                    # text_auto='.2f',\n",
    "                    color_discrete_map={\n",
    "                        \"Rare\": \"#545454\",\n",
    "                        \"Phaeocystis\": \"#F26B38\",\n",
    "                    },\n",
    "                    # Specify all the months that need to be included, \n",
    "                    # even if no sample has been taken\n",
    "                    category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                                        \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                                        \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                                        \"June_2021\", \"July_2021\"],\n",
    "                                    #\"genus\": [\"Rare\", \"Alexandrium\", \"Amphidinium\", \"Noctiluca\", \n",
    "                                    #          \"Symbiodinium\", \"Ceratium\", \"Tripos\", \"Kryptoperidinium\", \n",
    "                                    #          \"Oxyrrhis\", \"Karenia\", \"Durinskia\", \"Heterocapsa\", \"Scripsiella\"],\n",
    "                                    \"Genus_UniEuk\": ['Rare', 'Phaeocystis', 'Calcidiscus', 'Chrysochromulina', \n",
    "                                                     'Prymnesium', 'Scyphosphaera', 'Isochrysis', 'Emiliania', \n",
    "                                                     'Dicrateria', 'Haptolina', 'Coccolithus', 'Gephyrocapsa']}\n",
    "                    )\n",
    "\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "            size=8,  # Set the font size\n",
    "            color=\"#000000\"  # Set the font color\n",
    "        ),\n",
    "        autosize=False,\n",
    "        width= 8.5 * pixels_per_cm,\n",
    "        height= 7.5 * pixels_per_cm,\n",
    "        margin=dict( # Set the margins\n",
    "            l=0,  # Left margin\n",
    "            r=25,  # Right margin\n",
    "            b=25,  # Bottom margin\n",
    "            t=25  # Top margin\n",
    "        ),\n",
    "        xaxis_title_text='% TPM of total sum',\n",
    "        yaxis_title_text=None,\n",
    "        title_text=station\n",
    "    )\n",
    "\n",
    "    fig.show()\n",
    "\n",
    "    # Save figure as png\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophyte_relative_expression_per_month_{}_at_{}.png\".format(tax_level, station), scale=1)\n",
    "    #fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophyte_relative_expression_per_month_{}_at_{}.svg\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophyte_relative_expression_per_month_{}_at_{}_eukprot.png\".format(tax_level, station), scale=1)\n",
    "    fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophyte_relative_expression_per_month_{}_at_{}_eukprot.svg\".format(tax_level, station), scale=1)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Total TPM of prymnesiophytes per month"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'species'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "aggregation_level = 'month'\n",
    "\n",
    "#data_prymnesiophytes = data[(data['class'] == 'Prymnesiophyceae') & (data['genus'] == 'Phaeocystis') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_prymnesiophytes = data[(data['Taxogroup2_UniEuk'] == 'Prymnesiophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "# Remove data without a genus or with NaN values\n",
    "data_prymnesiophytes[tax_level].replace('', np.nan, inplace=True)\n",
    "data_prymnesiophytes.dropna(subset=tax_level, inplace=True)\n",
    "\n",
    "# Plot the total TPM per month\n",
    "fig = px.histogram(data_prymnesiophytes.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = tax_level,\n",
    "            color_discrete_map={\n",
    "                    \"Rare\": \"#545454\",\n",
    "                    \"Phaeocystis\": \"#F26B38\",\n",
    "                },\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"],\n",
    "                \"Genus_UniEuk\": [\"Rare\", \"Phaeocystis\"]},)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 11.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total prymnesiophyte TPM per month',\n",
    "    yaxis_title_text=None,\n",
    "    # Set range of x-axis\n",
    "    xaxis_range=[0, 550000],\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_total_TPM_per_month_{tax_level}.svg\", scale=1)\n",
    "fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_total_TPM_per_month_{tax_level}_eukprot.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Log version for clarity\n",
    "data_prymnesiophytes_log = data_prymnesiophytes.copy()\n",
    "data_prymnesiophytes_log[\"TPM\"] = np.log10(data_prymnesiophytes_log[\"TPM\"] + 1)\n",
    "\n",
    "# Plot the total TPM per month\n",
    "fig = px.histogram(data_prymnesiophytes_log.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = tax_level,\n",
    "            color_discrete_map={\n",
    "                    \"Rare\": \"#545454\",\n",
    "                    \"Phaeocystis\": \"#F26B38\",\n",
    "                },\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"],\n",
    "                \"Genus_UniEuk\": [\"Rare\", \"Phaeocystis\"]},)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='log(Total prymnesiophyte TPM per month)',\n",
    "    yaxis_title_text=None,\n",
    "    # Set range of x-axis\n",
    "#    xaxis_range=[0, 550000],\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_total_TPM_per_month_{tax_level}_log.svg\", scale=1)\n",
    "fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_total_TPM_per_month_{tax_level}_eukprot_log.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the total TPM per month per station\n",
    "fig = px.histogram(data_prymnesiophytes.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = 'station',\n",
    "            color_discrete_map={\n",
    "                \"ZG02\": \"#8c613c\",\n",
    "                \"120\": \"#956cb4\",\n",
    "                \"330\": \"#ee854a\",\n",
    "                \"130\": \"#4878d0\",\n",
    "                \"780\": \"#d65f5f\",\n",
    "                \"700\": \"#6acc64\"},\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"]},)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total prymnesiophyte TPM per month',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_total_TPM_per_month_station.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Log version for clarity\n",
    "data_prymnesiophytes_log = data_prymnesiophytes.copy()\n",
    "data_prymnesiophytes_log['TPM'] = np.log10(data_prymnesiophytes_log['TPM'])\n",
    "\n",
    "# Plot the total TPM per month\n",
    "fig = px.histogram(data_prymnesiophytes_log.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = 'station',\n",
    "            color_discrete_map={\n",
    "                \"ZG02\": \"#8c613c\",\n",
    "                \"120\": \"#956cb4\",\n",
    "                \"330\": \"#ee854a\",\n",
    "                \"130\": \"#4878d0\",\n",
    "                \"780\": \"#d65f5f\",\n",
    "                \"700\": \"#6acc64\"},\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"]},)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total prymnesiophyte TPM per month',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_total_TPM_per_month_station_log.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's repeat these last plots but for Phaeocystis species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'species'\n",
    "tax_level = 'Name_to_Use'\n",
    "aggregation_level = 'month'\n",
    "\n",
    "#data_prymnesiophytes = data[(data['class'] == 'Prymnesiophyceae') & (data['genus'] == 'Phaeocystis') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "data_prymnesiophytes = data[(data['Taxogroup2_UniEuk'] == 'Prymnesiophyceae') & (data['Genus_UniEuk'] == 'Phaeocystis') & (data['p_ident'] >= 0.95)][[aggregation_level, tax_level, 'TPM']]\n",
    "\n",
    "# Plot the total TPM per month\n",
    "fig = px.histogram(data_prymnesiophytes.sort_values(\"month\", ascending=False),\n",
    "             x = \"TPM\",\n",
    "             y = \"month\",\n",
    "             color = tax_level,\n",
    "            color_discrete_map={\n",
    "                    \"Rare\": \"#545454\",\n",
    "                    \"Phaeocystis\": \"#F26B38\",\n",
    "                },\n",
    "             orientation='h',\n",
    "            category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"],\n",
    "                \"Genus_UniEuk\": [\"Rare\", \"Phaeocystis\"]},)\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8,  # Set the font size\n",
    "        color=\"#000000\"  # Set the font color\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 11.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Total phaeocystis TPM per month',\n",
    "    yaxis_title_text=None,\n",
    "    # Set range of x-axis\n",
    "    xaxis_range=[0, 550000],\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_phaeocystis_total_TPM_per_month_{tax_level}.svg\", scale=1)\n",
    "fig.write_image(f\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_phaeocystis_total_TPM_per_month_{tax_level}_eukprot.svg\", scale=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Biodiversity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter the data\n",
    "## Get the diatoms\n",
    "# tax_level = 'genus'\n",
    "tax_level = 'Name_to_Use'\n",
    "aggregation_level = 'sample'\n",
    "\n",
    "## Since we'll be looking at the relative abundance of different diatom genera, we can only include reads that are annotated to a genus with a sufficient % sequence identity\n",
    "data_prymnesiophytes = data[(data['Taxogroup2_UniEuk'] == 'Prymnesiophyceae') & (data['p_ident'] >= 0.9)][[aggregation_level, tax_level, 'TPM']]\n",
    "# Remove transcripts below a certain TPM threshold\n",
    "data_prymnesiophytes = data_prymnesiophytes[data_prymnesiophytes['TPM'] > 1]\n",
    "\n",
    "# First, I only want to include transcriptome bins that have at least 100 non-zero expressed transcripts in one of the samples\n",
    "# Group by month and genus, sum TPM\n",
    "data_prymnesiophytes = data_prymnesiophytes.groupby([aggregation_level, tax_level]).count().reset_index()\n",
    "# Now transform the data to the wide format\n",
    "data_prymnesiophytes = data_prymnesiophytes.pivot(index=tax_level, columns=aggregation_level, values='TPM')\n",
    "# Set NaN values to 0\n",
    "data_prymnesiophytes = data_prymnesiophytes.fillna(0)\n",
    "\n",
    "# Visualize the data\n",
    "data_prymnesiophytes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_prymnesiophytes[data_prymnesiophytes > 100].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check how many diatom genera exist per month that have > 100 non-zero TPM transcripts\n",
    "## Per sample, count the number of genera that have > 100 TPM\n",
    "taxonomic_bin_abundance = data_prymnesiophytes[data_prymnesiophytes > 100].count(axis=0).reset_index()\n",
    "taxonomic_bin_abundance.columns = ['sample', 'num_species']\n",
    "\n",
    "## Add metadata\n",
    "taxonomic_bin_abundance = taxonomic_bin_abundance.merge(meta[['month']], left_on='sample', right_index=True, how='left')\n",
    "\n",
    "# Plot the number of diatom genera per month that have > 100 TPM\n",
    "\n",
    "fig = px.box(taxonomic_bin_abundance, x='month', y='num_species',\n",
    "             category_orders={'month': ['July_2020', \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "               \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "               \"May_2021\", \"June_2021\", \"July_2021\"]})\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# Save the figure\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/num_prymnesiophyte_species_per_month.png\", width = 3.5, height = 2, scale=1, format='svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the genera that have > 100 TPM in at least one sample\n",
    "taxonomic_bin_abundance = data_prymnesiophytes[data_prymnesiophytes > 100].count(axis=1)\n",
    "taxonomic_bin_abundance.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot distribution of most abundant diatom genera\n",
    "taxonomic_bin_abundance.sort_values(ascending=False).plot(kind='bar', figsize=(10, 5))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trophic feeding mode of prymnesiophytes\n",
    "I've predicted the trophic feeding mode of several species using the machine learning model of the paper by [Lambert & Groussman, 2022](https://doi.org/10.1073/pnas.2100916119), see [here](./trophic_mode_prediction.ipynb). We'll use the data to see how the trophic feeding mode of prymnesiophytes changes over time, same as for the diatoms and dinoflagellates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the data\n",
    "#trophic_predictions = pd.read_csv('../../data/analysis/phylodb_trophic_mode_predictions.csv')\n",
    "trophic_predictions = pd.read_csv('../../data/analysis/eukprot_trophic_mode_predictions.csv')\n",
    "\n",
    "# Only retain the diatoms\n",
    "#trophic_predictions = trophic_predictions[trophic_predictions['class'] == 'Prymnesiophyceae']\n",
    "trophic_predictions = trophic_predictions[trophic_predictions['Taxogroup2_UniEuk'] == 'Prymnesiophyceae']\n",
    "print(len(trophic_predictions))\n",
    "trophic_predictions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the relative abundance of the three trophic modes per month\n",
    "# Calculate relative abundance of each prediction per month\n",
    "rel_abundance = (trophic_predictions[['month', 'station', 'prediction']].groupby(['month', 'prediction'])\n",
    "                       .size()\n",
    "                       .groupby(level=0).apply(lambda x: 100 * x / x.sum())\n",
    "                       .reset_index(name='relative abundance'))\n",
    "# Set the color palette\n",
    "color_map = {\n",
    "    'Phot': 'green',\n",
    "    'Mix': 'black',\n",
    "    'Het': 'red'\n",
    "}\n",
    "fig = px.histogram(rel_abundance.sort_values(\"month\", ascending=False),\n",
    "             x = \"relative abundance\",\n",
    "             y = \"month\",\n",
    "             color = \"prediction\",\n",
    "             orientation='h',\n",
    "             color_discrete_map=color_map,\n",
    "             category_orders={\"month\": [\"July_2020\", \"August_2020\", \"September_2020\",\n",
    "                \"November_2020\", \"December_2020\", \"January_2021\",\n",
    "                \"February_2021\", \"April_2021\", \"May_2021\",\n",
    "                \"June_2021\", \"July_2021\"]})\n",
    "\n",
    "pixels_per_cm = 37.7952755906\n",
    "\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Times New Roman, serif\",  # Set the font family to Times New Roman\n",
    "        size=8  # Set the font size\n",
    "    ),\n",
    "    autosize=False,\n",
    "    width= 8.5 * pixels_per_cm,\n",
    "    height= 7.5 * pixels_per_cm,\n",
    "    margin=dict( # Set the margins\n",
    "        l=0,  # Left margin\n",
    "        r=25,  # Right margin\n",
    "        b=25,  # Bottom margin\n",
    "        t=25  # Top margin\n",
    "    ),\n",
    "    xaxis_title_text='Ratio of predicted trophic modes (%)',\n",
    "    yaxis_title_text=None,\n",
    ")\n",
    "\n",
    "fig.show()\n",
    "\n",
    "# save figure as svg\n",
    "#fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophyte_trophic_mode_relative_abundance_phylodb.svg\")\n",
    "fig.write_image(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophyte_trophic_mode_relative_abundance_eukprot.svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tax_level = 'genus'\n",
    "tax_level = 'Genus_UniEuk'\n",
    "\n",
    "# Extract the relevant columns\n",
    "df = trophic_predictions[[tax_level, 'month', 'prediction']]\n",
    "# Convert 'prediction' into numerical categories\n",
    "df['prediction'] = df['prediction'].map({'Phot':1, 'Mix':2, 'Het':3})\n",
    "\n",
    "# Pivot the DataFrame to use for heatmap\n",
    "df_pivot = df.pivot_table(index=tax_level, columns='month', values='prediction', \n",
    "                          aggfunc=lambda x: int(stats.mode(x[x!=0])[0][0] if any(x!=0) else 0))\n",
    "\n",
    "# Fill NA with a specific category (e.g., 0 for 'no prediction')\n",
    "df_pivot = df_pivot.fillna(0)\n",
    "\n",
    "# Ensure the values are integer\n",
    "df_pivot = df_pivot.astype(int)\n",
    "\n",
    "# Set the order of the months\n",
    "month_order = [\"July_2020\", \"August_2020\", \"September_2020\", \"November_2020\", \n",
    "                \"December_2020\", \"January_2021\", \"February_2021\", \"April_2021\", \n",
    "                \"May_2021\", \"June_2021\", \"July_2021\"]\n",
    "df_pivot = df_pivot.reindex(month_order, axis=1)\n",
    "\n",
    "# Set the order of the months\n",
    "df_pivot = df_pivot.reindex(month_order, axis=1)\n",
    "\n",
    "# Set the order of the genera to be the same as in the above relative abundance plot\n",
    "df_pivot = df_pivot.reindex(prymnesiophyte_legend_order, axis=0)\n",
    "\n",
    "# Create a color map\n",
    "cmap = mcolors.ListedColormap(['lightgrey', 'green', 'yellow', 'red'])\n",
    "\n",
    "# Create the heatmap with the color map\n",
    "plt.figure(figsize=(4.5, 4))\n",
    "sns.heatmap(df_pivot, cmap=cmap, annot=False, cbar=False, vmin=0, vmax=3)\n",
    "\n",
    "# Change layout to match the relative abundance plot\n",
    "plt.tight_layout()\n",
    "\n",
    "# Remove the x and  y-axis label\n",
    "plt.xlabel(None)\n",
    "plt.ylabel(None)\n",
    "\n",
    "# Font size should be 8, and font family Times New Roman\n",
    "plt.rcParams.update({'font.size': 8, 'font.family': 'Times New Roman'})\n",
    "\n",
    "# Save figure as svg\n",
    "#plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_trophic_mode_genus_consensus_heatmap.svg\", format='svg', dpi=600)\n",
    "plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_trophic_mode_genus_consensus_heatmap_eukprot.svg\", format='svg', dpi=600)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repeat for species level\n",
    "# Extract the relevant columns\n",
    "df = trophic_predictions[['species', 'month', 'prediction']]\n",
    "# Convert 'prediction' into numerical categories\n",
    "df['prediction'] = df['prediction'].map({'Phot':1, 'Mix':2, 'Het':3})\n",
    "\n",
    "# Pivot the DataFrame to use for heatmap\n",
    "df_pivot = df.pivot_table(index='species', columns='month', values='prediction', \n",
    "                          aggfunc=lambda x: int(stats.mode(x[x!=0])[0][0] if any(x!=0) else 0))\n",
    "\n",
    "# Fill NA with a specific category (e.g., 0 for 'no prediction')\n",
    "df_pivot = df_pivot.fillna(0)\n",
    "\n",
    "# Ensure the values are integer\n",
    "df_pivot = df_pivot.astype(int)\n",
    "\n",
    "# Set the order of the months\n",
    "df_pivot = df_pivot.reindex(month_order, axis=1)\n",
    "\n",
    "# Create a color map\n",
    "cmap = mcolors.ListedColormap(['lightgrey', 'green', 'yellow', 'red'])\n",
    "\n",
    "# Create the heatmap with the color map\n",
    "plt.figure(figsize=(4.5, 4))\n",
    "sns.heatmap(df_pivot, cmap=cmap, annot=False, cbar=False, vmin=0, vmax=3)\n",
    "\n",
    "# Change layout to match the relative abundance plot\n",
    "plt.tight_layout()\n",
    "\n",
    "# Remove the x and  y-axis label\n",
    "plt.xlabel(None)\n",
    "plt.ylabel(None)\n",
    "\n",
    "# Font size should be 8, and font family Times New Roman\n",
    "plt.rcParams.update({'font.size': 8, 'font.family': 'Times New Roman'})\n",
    "\n",
    "# Save figure as svg\n",
    "#plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_trophic_mode_species_consensus_heatmap.svg\", format='svg', dpi=600)\n",
    "plt.savefig(\"../../figures/diatoms_vs_dinoflagellates/prymnesiophytes_trophic_mode_species_consensus_heatmap_eukprot.svg\", format='svg', dpi=600)\n",
    "\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b0fa6594d8f4cbf19f97940f81e996739fb7646882a419484c72d19e05852a7e"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
