#!/usr/bin/bash -l

#PBS -l nodes=1:ppn=30
#PBS -l walltime=24:00:00
#PBS -l mem=100gb

source ../config.sh

# Align individual BPNS metatranscriptomic assemblies (mostly from large planktonic size fractions) on every MAG assembly 

unset OMP_PROC_BIND

module load MMseqs2

# Mapping of assembled transcripts on the MATOU DB
# Downloaded the MATOU DB from https://www.genoscope.cns.fr/tara/ and store in reference db folder

## Turn the MATOU unigenes into a mmseqs DB
mmseqs createdb ${MATOU} ${MATOU_mmseqsDB}

## Turn the query metatranscriptome into a mmseqs DB
mmseqs createdb ${assembly} ${metatranscriptome_mmseqsDB}

## Query metatranscriptome against reference
mmseqs search ${metatranscriptome_mmseqsDB} ${MATOU_mmseqsDB} ${MATOU_output} ${tmp} -s 6 --search-type 3
 -s: sensitivity, set to be a bit more sensitive
# search-type 3: nucleotide search

# Convert to blast tab formatted file
mmseqs convertalis ${metatranscriptome_mmseqsDB} ${MATOU_mmseqsDB} ${MATOU_output} ${MATOU_output}.m8

# Extract first hit
mmseqs filterdb ${MATOU_output} ${MATOU_output}.firsthit --extract-lines 1
mmseqs convertalis ${metatranscriptome_mmseqsDB} ${MATOU_mmseqsDB} ${MATOU_output}.firsthit ${MATOU_output}.firsthit.m8

# Extract first hit alignments with  a sequence ID > 95%
mmseqs filterdb ${MATOU_output}.firsthit ${MATOU_output}.firsthit.95plus_alnscore \
    --comparison-operator ge --comparison-value 0.95 --filter-column 3

mmseqs convertalis ${metatranscriptome_mmseqsDB} ${MATOU_mmseqsDB} ${MATOU_output}.firsthit.95plus_alnscore \
   ${MATOU_output}.firsthit.95plus_alnscore.m8``